<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>io Module &mdash; mapmuts 1.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="mapmuts 1.1 documentation" href="index.html" />
    <link rel="up" title="Python modules" href="pythonapi.html" />
    <link rel="next" title="stats Module" href="stats_module.html" />
    <link rel="prev" title="sequtils Module" href="sequtils_module.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="stats_module.html" title="stats Module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="sequtils_module.html" title="sequtils Module"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">mapmuts 1.1 documentation</a> &raquo;</li>
          <li><a href="pythonapi.html" accesskey="U">Python modules</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-src.io">
<span id="io-module"></span><h1><tt class="xref py py-mod docutils literal"><span class="pre">io</span></tt> Module<a class="headerlink" href="#module-src.io" title="Permalink to this headline">¶</a></h1>
<p>Performs input/output operations for <tt class="docutils literal"><span class="pre">mapmuts</span></tt> package.</p>
<p>Written by Jesse Bloom</p>
<div class="section" id="list-of-functions">
<h2>List of functions<a class="headerlink" href="#list-of-functions" title="Permalink to this headline">¶</a></h2>
<p><cite>IteratePairedFASTQ</cite> : iterates over FASTQ files for paired-end reads.</p>
<p><cite>IterateAlignmentFile</cite> : iterates over files written by <cite>main.MakeAlignments</cite>.</p>
<p><cite>WriteInsertLengths</cite> : writes distribution of insert lengths.</p>
<p><cite>ReadInsertLengths</cite> : reads data written by <cite>WriteInsertLengths</cite>.</p>
<p><cite>WriteAlignmentStatistics</cite> : writes summary of alignment statistics.</p>
<p><cite>ReadAlignmentStatistics</cite> : reads data written by <cite>WriteAlignmentStatistics</cite>.</p>
<p><cite>WriteReadMismatches</cite> : writes distribution of mismatches along read.</p>
<p><cite>ReadReadMismatches</cite> : reads data written by <cite>WriteReadMismatches</cite>.</p>
<p><cite>WriteNTCounts</cite> : writes counts of nucleotide identities.</p>
<p><cite>ReadNTCounts</cite> : reads output of <cite>WriteNTCounts</cite>.</p>
<p><cite>WriteCodonCounts</cite> : writes counts of codon identities.</p>
<p><cite>ReadCodonCounts</cite> : reads output of <cite>WriteCodonCounts</cite>.</p>
<p><cite>ReadEnrichmentRatios</cite> : reads enrichment ratios from <tt class="docutils literal"><span class="pre">mapmuts_inferenrichment.py</span></tt></p>
<p><cite>ReadEntropyAndEquilFreqs</cite> : reads entropies / equilibrium freqs from 
<tt class="docutils literal"><span class="pre">mapmuts_inferenrichment.py</span></tt></p>
<p><cite>WriteAAsFromCodonCounts</cite> : reads files created by <cite>WriteCodonCounts</cite>
and writes the amino acid frequencies.</p>
<p><cite>ParseInfile</cite> : reads key / value pairs from an input file.</p>
<p><cite>ParseBoolValue</cite> : reads Boolean values from input dictionary.</p>
<p><cite>ParseIntValue</cite> : reads integer values from input dictionary.</p>
<p><cite>ParseFloatValue</cite> : reads float values from input dictionary.</p>
<p><cite>ParseStringValue</cite> : reads string values from input dictionary.</p>
<p><cite>ParseSeqValue</cite> : reads sequence from file specified by input dictionary.</p>
<p><cite>ParseFileList</cite> : reads file list specified by input dictionary.</p>
<p><cite>AllNTs</cite> : lists nucleotide codes accepted by <cite>WriteNTCounts</cite> / <cite>ReadNTCounts</cite>.</p>
<p><cite>AllCodons</cite> : list codon codes accepted by <cite>WriteCodonCounts</cite> / <cite>ReadCodonCounts</cite>.</p>
<p><cite>PrintVersions</cite> : prints version numbers of <em>mapmuts</em> and associated packages.</p>
</div>
<div class="section" id="documentation-for-individual-functions">
<h2>Documentation for individual functions<a class="headerlink" href="#documentation-for-individual-functions" title="Permalink to this headline">¶</a></h2>
<p>Documentation for individual functions is provided in their definitions below.</p>
<dl class="function">
<dt id="src.io.AllCodons">
<tt class="descclassname">src.io.</tt><tt class="descname">AllCodons</tt><big>(</big><big>)</big><a class="headerlink" href="#src.io.AllCodons" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all codon codes for <cite>ReadCodonCounts</cite> / <cite>WriteCodonCounts</cite>.</p>
<p>Returns a list of all codon codes in the scheme used by
<cite>ReadCodonCounts</cite> / <cite>WriteCodonCounts</cite>. The order of entries in this list
is important, and should be preserved to maintain compatibility
for reading/writing files.</p>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">AllCodons</span><span class="p">()[</span> <span class="p">:</span> <span class="mi">9</span><span class="p">]</span>
<span class="go">[&#39;AAA&#39;, &#39;AAT&#39;, &#39;AAC&#39;, &#39;AAG&#39;, &#39;AAa&#39;, &#39;AAt&#39;, &#39;AAc&#39;, &#39;AAg&#39;, &#39;AAN&#39;]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">AllCodons</span><span class="p">())</span> <span class="o">==</span> <span class="mi">9</span><span class="o">**</span><span class="mi">3</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.AllNTs">
<tt class="descclassname">src.io.</tt><tt class="descname">AllNTs</tt><big>(</big><big>)</big><a class="headerlink" href="#src.io.AllNTs" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all nucleotide codes for <cite>ReadNTCounts</cite> / <cite>WriteNTCounts</cite>.</p>
<p>Returns a list of all nucleotide codes in the scheme used by
<cite>ReadNTCounts</cite> / <cite>WriteNTCounts</cite>. The order of entries in this list
is important, and should be preserved to maintain compatibility
for reading/writing files.</p>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">AllNTs</span><span class="p">()</span>
<span class="go">[&#39;A&#39;, &#39;T&#39;, &#39;C&#39;, &#39;G&#39;, &#39;a&#39;, &#39;t&#39;, &#39;c&#39;, &#39;g&#39;, &#39;N&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.IterateAlignmentFile">
<tt class="descclassname">src.io.</tt><tt class="descname">IterateAlignmentFile</tt><big>(</big><em>filename</em>, <em>gzipped=True</em>, <em>usegzip=False</em><big>)</big><a class="headerlink" href="#src.io.IterateAlignmentFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates over alignment file written by main.MakeAlignments</p>
<p><cite>filename</cite> : name of a file of the type written by <cite>main.MakeAlignments</cite>
as the <tt class="docutils literal"><span class="pre">*_alignments.txt.gz</span></tt> file. Each entry
in these alignment files consists of five lines, giving the
read name, gene, read 1, read 2, blank line, as in:</p>
<div class="highlight-python"><div class="highlight"><pre>@DH1DQQN1:241:C1433ACXX:2:1101:1623:2365
14 ATAGATACTAGAT 26
 1 .....C.....T. 13
13 .....C.......  1
</pre></div>
</div>
<p><cite>gzipped</cite> : Boolean switch specifying whether the alignment
file is gzipped. In general, this will be <cite>True</cite> since
<cite>main.MakeAlignments</cite> writes gzipped files. Default
value is <cite>True</cite>.</p>
<p><cite>usegzip</cite> : a Boolean switch that is meaningful only if <cite>gzipped</cite> 
is <cite>True</cite>. This specifies that we use the Python <tt class="docutils literal"><span class="pre">gzip</span></tt> module to 
unzip the reads. Otherwise we use the system <tt class="docutils literal"><span class="pre">gzip</span> <span class="pre">-cd</span></tt> command,
which may be faster. You should definitely specify
<cite>usegzip</cite> to <cite>True</cite> if you are only reading a few lines, as with
<cite>usegzip</cite> as <cite>False</cite> the entire file is unzipped before anything happens.
<cite>False</cite> by default.</p>
<p>The function will iterate over all alignment entries in
filename. Each iteration returns the following tuple:
<cite>(gstart, gend, gseq, r1start, r1end, r1seq, r2start, r2end, r2seq)</cite>
The tuple entries have the following meanings:</p>
<blockquote>
<div><p><cite>gstart</cite> : index of first gene nt; 1, 2, ... numbering.</p>
<p><cite>gend</cite> : index last gene nt; 1, 2, ... numbering.</p>
<p><cite>gseq</cite> : aligned gene nucleotides.</p>
<p><cite>r1start</cite> -&gt; index of first R1 nt; 1, 2, ... numbering.</p>
<p><cite>r1end</cite> : index of last R1 nt; 1, 2, ... numbering.</p>
<p><cite>r1seq</cite> : aligned R1 nucleotides, with dots at identities.</p>
<p><cite>r2start</cite> : index of first R2 nt; 1, 2, ... numbering.</p>
<p><cite>r2end</cite> : index of last R2 nt; 1, 2, ... numbering.</p>
<p><cite>r2seq</cite> : aligned R2 nucleotides, with dots at identities.</p>
</div></blockquote>
<p>Note that the <cite>r1</cite> and <cite>r2</cite> indices are for the orientation in
which the read aligns, so either <cite>r1start &lt;= r1end</cite> or
<cite>r2start &lt;= r2end</cite>.</p>
<p>For example, the alignment shown above under the filename
option description would return:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">26</span><span class="p">,</span> <span class="s">&#39;ATAGATACTAGAT&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="s">&#39;.....C.....T.&#39;</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s">&#39;.....C.......&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.IteratePairedFASTQ">
<tt class="descclassname">src.io.</tt><tt class="descname">IteratePairedFASTQ</tt><big>(</big><em>r1files</em>, <em>r2files</em>, <em>gzipped</em>, <em>applyfilter</em>, <em>usegzip=False</em><big>)</big><a class="headerlink" href="#src.io.IteratePairedFASTQ" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates over FASTQ file pairs for paired-end sequencing reads.</p>
<p>This function has been tested for its ability to read FASTQ files
generated by the Illumina Casava 1.8 pipeline for paired-end reads.</p>
<p>CALLING VARIABLES:</p>
<p><cite>r1files</cite> : list of file name(s) containing the R1 reads.</p>
<p><cite>r2files</cite> : list of file name(s) containing the R2 reads. The files
specified here must specify the exact same number of reads as
those in <cite>r1files</cite>, and they must be in the same sequence (i.e.
the first R1 read found going through <cite>r1files</cite> in order must
match the first R2 read found going through <cite>r2files</cite> in order,
etc).</p>
<p><cite>gzipped</cite> : Boolean switch specifying whether the FASTQ files
specified by <cite>r1files</cite> and <cite>r2files</cite> are gzipped. If set to <cite>True</cite>,
they are gzipped. This iterator will then read them without
unzipping the files.</p>
<p><cite>applyfilter</cite> : Boolean switch specifying whether we remove read
pairs in which one or more of the reads failed the Illumina
chastity filter. If <cite>True</cite>, the iterator simply returns <cite>False</cite>
each time it encounters a read pair where the chastity filter is
failed. The chastity filter is indicated by Y or N in the
sequence header &#8211; values of Y indicate a failed filter.</p>
<p><cite>usegzip</cite> : Boolean switch that is meaningful only if <cite>gzipped</cite>
is <cite>True</cite>. This specifies that we use the Python <tt class="docutils literal"><span class="pre">gzip</span></tt> module to 
unzip the reads. Otherwise we use the system <tt class="docutils literal"><span class="pre">gzip</span> <span class="pre">-cd</span> <span class="pre">command</span></tt>,
which may be faster. You should definitely specify
<cite>usegzip</cite> to <cite>True</cite> if you are only reading a few lines, as with
<cite>usegzip</cite> set to <cite>False</cite> the entire file is unzipped before anything happens.
<cite>False</cite> by default.</p>
<p>RESULT OF THIS FUNCTION:</p>
<p>The function will iterate over all read pairs in the indicated
files. For each iteration, it will return:</p>
<ul>
<li><p class="first">The return variable is <cite>False</cite> if <tt class="docutils literal"><span class="pre">applyfilter</span> <span class="pre">==</span> <span class="pre">True</span></tt> and one of the
reads in the pair failed the Illumina chastity filter.</p>
</li>
<li><p class="first">Otherwise the
return variable is the following tuple:
<cite>(name, r1, r2, q1avg, q2avg)</cite> where:</p>
<p><cite>name</cite> : name of the read pair (a string)</p>
<p><cite>r1</cite> : sequence of the first read R1 (a string)</p>
<p><cite>r2</cite> : sequence of the second read R2 (a string)</p>
<p><cite>q1avg</cite> : the mean Q-score for the R1 read (a float)</p>
<p><cite>q2avg</cite> : the mean Q-score for the R2 read (a float)</p>
</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="src.io.ParseBoolValue">
<tt class="descclassname">src.io.</tt><tt class="descname">ParseBoolValue</tt><big>(</big><em>d</em>, <em>key</em><big>)</big><a class="headerlink" href="#src.io.ParseBoolValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets Boolean argument from input dictionary.</p>
<p><cite>d</cite> : a dictionary of key/value string pairs as returned by
<cite>ParseInfile</cite>.</p>
<dl class="docutils">
<dt><cite>key</cite> <span class="classifier-delimiter">:</span> <span class="classifier">a string in this dictionary that specifies a value that</span></dt>
<dd>should be &#8216;True&#8217; or &#8216;False&#8217;.</dd>
</dl>
<p>Returns the Boolean truth value specified by <cite>key</cite>. Raises a <cite>ValueError</cite>
if <cite>key</cite> does not exist in <cite>d</cite>, and a <cite>ValueError</cite> if it specifies
something other than <cite>True</cite> or <cite>False</cite>.</p>
<p>Here are some examples:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;gzipped&#39;</span><span class="p">:</span><span class="s">&#39;True&#39;</span><span class="p">,</span> <span class="s">&#39;applyfilter&#39;</span><span class="p">:</span><span class="s">&#39;False&#39;</span><span class="p">,</span> <span class="s">&#39;a1&#39;</span><span class="p">:</span><span class="s">&#39;a1.txt&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ParseBoolValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;gzipped&#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ParseBoolValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;applyfilter&#39;</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ParseBoolValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;a1&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">value a1.txt for a1 is not True/False</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ParseBoolValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;otherkey&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">did not find a key for otherkey</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.ParseFileList">
<tt class="descclassname">src.io.</tt><tt class="descname">ParseFileList</tt><big>(</big><em>d</em>, <em>key</em><big>)</big><a class="headerlink" href="#src.io.ParseFileList" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets list of files from input dictionary.</p>
<p><cite>d</cite> : a dictionary of key/value string pairs as returned by
<cite>ParseInfile</cite>.</p>
<p><cite>key</cite> : a string in this dictionary that specifies a value that
should give a list of one or more file names.</p>
<p>Returns a list of the file names specified by <cite>key</cite>. If one or more of
these files does not exist, raises and <cite>IOError</cite>. If <cite>key</cite> does not
exist in <cite>d</cite>, raises an <cite>KeyError</cite>.</p>
</dd></dl>

<dl class="function">
<dt id="src.io.ParseFloatValue">
<tt class="descclassname">src.io.</tt><tt class="descname">ParseFloatValue</tt><big>(</big><em>d</em>, <em>key</em><big>)</big><a class="headerlink" href="#src.io.ParseFloatValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets float argument from input dictionary.</p>
<p><cite>d</cite> : a dictionary of key/value string pairs as returned by
<cite>ParseInfile</cite>.</p>
<p><cite>key</cite> : a string in this dictionary that specifies a value that
should be convertible to a float.</p>
<p>Returns the float specified by <cite>key</cite>. Raises a <cite>ValueError</cite>
if <cite>key</cite> does not exist in <cite>d</cite>, and a <cite>ValueError</cite> if it specifies
something other than an integer.</p>
<p>Here are some examples:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;maxn&#39;</span><span class="p">:</span><span class="s">&#39;2&#39;</span><span class="p">,</span> <span class="s">&#39;minq&#39;</span><span class="p">:</span><span class="s">&#39;20.5&#39;</span><span class="p">,</span> <span class="s">&#39;string&#39;</span><span class="p">:</span><span class="s">&#39;hello&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.1f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">ParseFloatValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;maxn&#39;</span><span class="p">)</span>
<span class="go">2.0</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&quot;</span><span class="si">%.1f</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">ParseFloatValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;minq&#39;</span><span class="p">)</span>
<span class="go">20.5</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ParseFloatValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;string&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">value hello for string is not float</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ParseFloatValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;otherkey&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">did not find a key for otherkey</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.ParseInfile">
<tt class="descclassname">src.io.</tt><tt class="descname">ParseInfile</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#src.io.ParseInfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads key / value pairs from an input file.</p>
<p><cite>f</cite> : should be a readable file-like object.</p>
<p>Starting at the current position in <cite>f</cite>, reads all remaining lines
until the end of the file-like object. Any blank line or line
beginning with a # character is disregarded (# indicates a comment
line). All other lines should contain exactly two entries, the first
being the key and the second being the value. The key is construed
to be the first word, and cannot have any spaces. The value is all
of the text that follows the key up to the newline.</p>
<p>The key / value pairs are returned in a
dictionary, as illustrated in the example below. If there is a
duplicate key, and exception is raised.</p>
<p>Example of successfully reading two key/value pairs:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# comment line followed by blank line</span><span class="se">\n\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;key1 first_value</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# now another key with two values</span><span class="se">\n</span><span class="s">key2 2 value2&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ParseInfile</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="o">==</span> <span class="p">{</span><span class="s">&#39;key1&#39;</span><span class="p">:</span><span class="s">&#39;first_value&#39;</span><span class="p">,</span> <span class="s">&#39;key2&#39;</span><span class="p">:</span><span class="s">&#39;2 value2&#39;</span><span class="p">}</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Example of duplicate key name:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# comment line followed by blank line</span><span class="se">\n\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;key1 first_value</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;# now another key</span><span class="se">\n</span><span class="s">key2 2&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\n</span><span class="s">key1 1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ParseInfile</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="o">==</span> <span class="p">{</span><span class="s">&#39;key1&#39;</span><span class="p">:</span><span class="s">&#39;first_value&#39;</span><span class="p">,</span> <span class="s">&#39;key2&#39;</span><span class="p">:</span><span class="s">&#39;2&#39;</span><span class="p">}</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError: duplicate key</span>: <span class="n">key1</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.ParseIntValue">
<tt class="descclassname">src.io.</tt><tt class="descname">ParseIntValue</tt><big>(</big><em>d</em>, <em>key</em><big>)</big><a class="headerlink" href="#src.io.ParseIntValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets integer argument from input dictionary.</p>
<p><cite>d</cite> : a dictionary of key/value string pairs as returned by
<cite>ParseInfile</cite>.</p>
<p><cite>key</cite> : a string in this dictionary that specifies a value that
should be an integer.</p>
<p>Returns the integer specified by <cite>key</cite>. Raises a <cite>ValueError</cite>
if <cite>key</cite> does not exist in <cite>d</cite>, and a <cite>ValueError</cite> if it specifies
something other than an integer.</p>
<p>Here are some examples:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;maxn&#39;</span><span class="p">:</span><span class="s">&#39;2&#39;</span><span class="p">,</span> <span class="s">&#39;minq&#39;</span><span class="p">:</span><span class="s">&#39;20.5&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ParseIntValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;maxn&#39;</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ParseIntValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;minq&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">value 20.5 for minq is not integer</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ParseIntValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;otherkey&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">did not find a key for otherkey</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.ParseSeqValue">
<tt class="descclassname">src.io.</tt><tt class="descname">ParseSeqValue</tt><big>(</big><em>d</em>, <em>key</em><big>)</big><a class="headerlink" href="#src.io.ParseSeqValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads sequence from FASTA file specified by input dictionary.</p>
<p><cite>d</cite> : a dictionary of key/value string pairs as returned by
<cite>ParseInfile</cite>.</p>
<p><cite>key</cite> : a string in this dictionary that specifies a value that is a
filename of a readable file that contains exactly one sequence
in FASTA format.</p>
<p>Returns a string corresponding to the sequence specified in the
FASTA file. Raises a <cite>ValueError</cite> if the filename does not exist, or
does not specify exactly one sequence.</p>
</dd></dl>

<dl class="function">
<dt id="src.io.ParseStringValue">
<tt class="descclassname">src.io.</tt><tt class="descname">ParseStringValue</tt><big>(</big><em>d</em>, <em>key</em><big>)</big><a class="headerlink" href="#src.io.ParseStringValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads string argument specified by input dictionary.</p>
<p><cite>d</cite> : a dictionary of the key/value string pairs as returned by
<cite>ParseInfile</cite>.</p>
<p><cite>key</cite> : a string in this dictionary that specifies a value that
is a string.</p>
<p>Returns string corresponding to <cite>key</cite>. Raises a <cite>ValueError</cite> if 
<cite>key</cite> does not exist as a key in <cite>d</cite>.</p>
<p>Here are some examples:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;outfileprefix&#39;</span><span class="p">:</span><span class="s">&#39;test_example&#39;</span><span class="p">,</span> <span class="s">&#39;samplename&#39;</span><span class="p">:</span><span class="s">&#39;test example&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ParseStringValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;outfileprefix&#39;</span><span class="p">)</span>
<span class="go">&#39;test_example&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ParseStringValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;samplename&#39;</span><span class="p">)</span>
<span class="go">&#39;test example&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ParseStringValue</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="s">&#39;otherkey&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">did not find a key for otherkey</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.PrintVersions">
<tt class="descclassname">src.io.</tt><tt class="descname">PrintVersions</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#src.io.PrintVersions" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints information about the versions of <tt class="docutils literal"><span class="pre">mapmuts</span></tt> and utilized packages.</p>
<p><em>f</em> is a writable file-like object (for example, could be <em>sys.stdout</em> if
you want to write to standard output).</p>
<p>This function writes to <em>f</em> a summary of the software versions. It
also writes the current time and platform. For the optional packages,
if any of the packages are not available then that is indicated. Prints
information about the following:</p>
<blockquote>
<div><ul class="simple">
<li>The current time / date (accessed via <em>time.asctime()</em>).</li>
<li>The current platform (accessed via <em>platform.platform()</em>).</li>
<li>The version of Python (accessed via <em>sys.version</em>).</li>
<li>The version of <tt class="docutils literal"><span class="pre">mapmuts</span></tt> (accessed via <em>mapmuts.__version__</em>).</li>
<li>The version of <tt class="docutils literal"><span class="pre">numpy</span></tt> (accessed via <em>numpy.__version__</em>).</li>
<li>The version of <tt class="docutils literal"><span class="pre">pymc</span></tt> (accessed via <em>pymc.__version__</em>).</li>
<li>The version of <tt class="docutils literal"><span class="pre">scipy</span></tt> (accessed via <em>scipy.__version__</em>).</li>
<li>The version of <tt class="docutils literal"><span class="pre">matplotlib</span></tt> (accessed via <em>matplotlib.__version__</em>).</li>
<li>The version of <tt class="docutils literal"><span class="pre">pyPdf</span></tt>: unfortunately the <tt class="docutils literal"><span class="pre">pyPdf</span></tt> package does
not contain a built-in version string, so we only print whether
<tt class="docutils literal"><span class="pre">pyPdf</span></tt> is installed at all.</li>
<li>The version of <tt class="docutils literal"><span class="pre">pdflatex</span></tt> (accessed via <em>mapmuts.latex.PdflatexAvailable()</em>).</li>
<li>The version of <tt class="docutils literal"><span class="pre">weblogo</span></tt> (acccessed via <em>mapmuts.latex.WebLogoAvailable()</em>).</li>
</ul>
</div></blockquote>
<p>Here is an example of the output of this function:</p>
<div class="highlight-python"><div class="highlight"><pre>****************************************************
Version information for mapmuts and associated programs.

Time and date: Thu Dec  5 16:40:30 2013

Platform: Darwin-10.8.0-x86_64-i386-64bit

Python version: 2.6.8 (unknown, Apr 14 2012, 04:15:37) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]

mapmuts version: 1.0

numpy version: 1.7.1

pymc version: 2.3

scipy version: 0.12.0

matplotlib version: 1.2.1

pyPdf version: pyPdf is available, but no version string accessible

pdflatex version: pdfTeX 3.1415926-1.40.11-2.2 (TeX Live 2010)
kpathsea version 6.0.0
Copyright 2010 Peter Breitenlohner (eTeX)/Han The Thanh (pdfTeX).
There is NO warranty.  Redistribution of this software is
covered by the terms of both the pdfTeX copyright and
the Lesser GNU General Public License.
For more information about these matters, see the file
named COPYING and the pdfTeX source.
Primary author of pdfTeX: Peter Breitenlohner (eTeX)/Han The Thanh (pdfTeX).
Compiled with libpng 1.2.40; using libpng 1.2.40
Compiled with zlib 1.2.3; using zlib 1.2.3
Compiled with xpdf version 3.02pl4

weblogo version: WebLogo 3.3 (2012-07-02)
****************************************************
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.ReadAlignmentStatistics">
<tt class="descclassname">src.io.</tt><tt class="descname">ReadAlignmentStatistics</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#src.io.ReadAlignmentStatistics" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads alignment statistics written by <cite>WriteAlignmentStatistics</cite>.</p>
<p><cite>f</cite> : a readable file-like object, opened to the position
where output was written by <cite>WriteAlignmentStatistics</cite>.</p>
<p>This function returns a dictionary <cite>s</cite> containing the statistics used
to call <cite>WriteAlignmentStatistics</cite> to generate the output in <cite>f</cite>.</p>
<p>Here is an example showing how <cite>ReadAlignmentStatistics</cite> recovers the
output written by <cite>WriteAlignmentStatistics</cite>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;nread&#39;</span><span class="p">:</span><span class="mi">110</span><span class="p">,</span> <span class="s">&#39;nexcessn&#39;</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span> <span class="s">&#39;nfiltered&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="s">&#39;nlowq&#39;</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span> <span class="s">&#39;nattempted&#39;</span><span class="p">:</span><span class="mi">90</span><span class="p">,</span> 
<span class="gp">... </span>   <span class="s">&#39;npaired&#39;</span><span class="p">:</span><span class="mi">65</span><span class="p">,</span> <span class="s">&#39;nalignedfullgene&#39;</span><span class="p">:</span><span class="mi">55</span><span class="p">,</span> <span class="s">&#39;nalignedgene&#39;</span><span class="p">:</span><span class="mi">50</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteAlignmentStatistics</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="n">ReadAlignmentStatistics</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.ReadCodonCounts">
<tt class="descclassname">src.io.</tt><tt class="descname">ReadCodonCounts</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#src.io.ReadCodonCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads counts of codon identities written by <cite>WriteCodonCounts</cite>.</p>
<p>This function reads the files written by <cite>WriteCodonCounts</cite> to
reconstruct the calling dictionary specifying the number
of time each codon identity is observed at a position.</p>
<p><cite>f</cite> : a readable file-like object of the type written by 
<cite>WriteNTCounts</cite>, open to the position where the codon
counts begin.</p>
<p>The returned variable is a dictionary that should be similar
to the <cite>codon_counts</cite> dictionary used to call <cite>WriteCodonCounts</cite>.</p>
</dd></dl>

<dl class="function">
<dt id="src.io.ReadEnrichmentRatios">
<tt class="descclassname">src.io.</tt><tt class="descname">ReadEnrichmentRatios</tt><big>(</big><em>infile</em><big>)</big><a class="headerlink" href="#src.io.ReadEnrichmentRatios" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads enrichment ratios written by <tt class="docutils literal"><span class="pre">mapmuts_inferenrichment.py</span></tt>.</p>
<p>The single calling argument <em>infile</em> should specify a 
<tt class="docutils literal"><span class="pre">*_enrichmentratios.txt</span></tt> file written by <tt class="docutils literal"><span class="pre">mapmuts_inferenrichment.py</span></tt>.
Specifically, these files have the following format:</p>
<div class="highlight-python"><div class="highlight"><pre>#MUTATION PHI PHI_HPD95_LOW PHI_HPD95_HIGH DIRECT_RATIO MUTDNA_COUNTS
A22N    0.030913    0.013112    0.049075    0.021807    364
A22Q    0.070328    0.043265    0.097744    0.062814    408
A22P    0.133431    0.103709    0.160012    0.203831    1128
A22S    1.936427    1.787140    2.098920    1.740187    1065
A22R    0.725581    0.653297    0.802220    0.747375    827
A22T    0.175550    0.141060    0.210771    0.243729    1140
A22W    0.141486    0.077679    0.211875    0.120190    121
</pre></div>
</div>
<p>where the first column gives the mutation, the second gives the enrichment
ratio (always a number &gt; 0),
the third and fourth give the 95% highest posterior density for the
enrichment ratio, the fifth column gives the directly estimated
enrichment ratio denotes the enrichment ratio (a number that can range
from minus infinity to infinity inclusive), and the fifth gives the number
of counts in the mutant library (a number &gt;= 0).</p>
<p>The returned variable is a dictionary keyed by mutation name. For each
key, the value is itself a dictionary with the string keys <em>PHI</em>,
<em>PHI_HPD95_LOW</em>, <em>PHI_HPD95_HIGH</em>, <em>DIRECT_RATIO</em>, and <em>MUTDNA_COUNTS</em>
with the values matching the numbers specified for that mutation in
<em>infile</em>. If a mutation is found multiple times in <em>infile</em>, and exception
is raised. For <em>DIRECT_RATIO</em>, if the values are <tt class="docutils literal"><span class="pre">-inf</span></tt> or <tt class="docutils literal"><span class="pre">inf</span></tt> then
the value is <em>-float(&#8216;inf&#8217;)</em> or <em>float(&#8216;inf&#8217;)</em>.</p>
</dd></dl>

<dl class="function">
<dt id="src.io.ReadEntropyAndEquilFreqs">
<tt class="descclassname">src.io.</tt><tt class="descname">ReadEntropyAndEquilFreqs</tt><big>(</big><em>infile</em><big>)</big><a class="headerlink" href="#src.io.ReadEntropyAndEquilFreqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads entropies and preferences written by <tt class="docutils literal"><span class="pre">mapmuts_inferenrichment.py</span></tt> and <tt class="docutils literal"><span class="pre">mapmuts_inferpreferences.py</span></tt>.</p>
<p>The single calling argument <em>infile</em> should specify a 
<tt class="docutils literal"><span class="pre">*_equilibriumfreqs.txt</span></tt> file written by <tt class="docutils literal"><span class="pre">mapmuts_inferenrichment.py</span></tt>
or an <tt class="docutils literal"><span class="pre">*_equilibriumpreferences.txt</span></tt> written by <tt class="docutils literal"><span class="pre">mapmuts_inferpreferences.py</span></tt>.</p>
<p>These files have the following format:</p>
<div class="highlight-python"><div class="highlight"><pre>#SITE WT_AA SITE_ENTROPY PI_A PI_C PI_D PI_E PI_F PI_G PI_H PI_I PI_K PI_L PI_M PI_N PI_P PI_Q PI_R PI_S PI_T PI_V PI_W PI_Y
1   M   3.952438    0.027276    0.057787    0.061477    0.0589700.058843    0.023888    0.050938    0.017797    0.062260    0.019259    0.216150    0.070343    0.029413    0.052597    0.018734    0.013009    0.025890    0.020677    0.058900    0.055793
2   A   2.362402    0.537118    0.012475    0.013213    0.0258260.014365    0.008897    0.005926    0.009136    0.013902    0.003019    0.191576    0.010007    0.001631    0.007791    0.003013    0.004130    0.005614    0.099395    0.020214    0.012752
3   S   2.745295    0.024799    0.002901    0.002167    0.0071370.311824    0.002504    0.022821    0.008849    0.004478    0.059771    0.015584    0.022394    0.005062    0.004390    0.006270    0.091925    0.014399    0.003761    0.048111    0.340854
</pre></div>
</div>
<p>The first column gives the residue number, the second gives the wildtype
residue, the third the site entropy, and the remainder the equilibrium
frequencies of each of the 20 amino acids. The columns can be delimited
by a single space or single tab.</p>
<p>There <strong>may</strong> be a final column <tt class="docutils literal"><span class="pre">PI_*</span></tt> giving the frequency for stop codons.
That column can either be present or absent &#8211; either one is OK.</p>
<p>The returned variable is a dictionary keyed by residue numbers, and with
the values for each residue number being dictionaries keyed by all of the
other 22 or 23 column labels (&#8216;WT_AA&#8217;, &#8216;SITE_ENTROPY&#8217;, &#8216;PI_A&#8217;, &#8216;PI_C&#8217;, ...)
and possibly &#8216;PI_*&#8217; if that is present in the header.</p>
</dd></dl>

<dl class="function">
<dt id="src.io.ReadInsertLengths">
<tt class="descclassname">src.io.</tt><tt class="descname">ReadInsertLengths</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#src.io.ReadInsertLengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads alignment statistics written by <cite>WriteInsertLengths</cite>.</p>
<p><cite>f</cite> : a readable file-like object, opened to the position
where output was written by <cite>WriteInsertLengths</cite>.</p>
<p>This function returns a dictionary <cite>d</cite> containing the statistics used
to call <cite>WriteInsertLengths</cite> to generate the output in <cite>f</cite>.</p>
<p>Here is an example showing how <cite>ReadInsertLengths</cite> recovers the
output written by <cite>WriteInsertLengths</cite>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="mi">20</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">21</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span> <span class="mi">22</span><span class="p">:</span><span class="mi">7</span><span class="p">,</span> <span class="mi">23</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">24</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span> <span class="mi">25</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">27</span><span class="p">:</span><span class="mi">1</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteInsertLengths</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">==</span> <span class="n">ReadInsertLengths</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.ReadNTCounts">
<tt class="descclassname">src.io.</tt><tt class="descname">ReadNTCounts</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#src.io.ReadNTCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads counts of nucleotide identities written by <cite>WriteNTCounts</cite>.</p>
<p>This function reads the files written by <cite>WriteNTCounts</cite> to
reconstruct the calling dictionary specifying the number
of time each nucleotide identity is observed at a position.</p>
<p><cite>f</cite> : a readable file-like object of the type written by 
<cite>WriteNTCounts</cite>, open to the position where the nucleotide
counts begin.</p>
<p>The returned variable is a dictionary that should be identical in
its contents to the <cite>nt_counts</cite> dictionary used to call <cite>WriteNTCounts</cite>.</p>
<p>Here is an example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">nt_counts</span> <span class="o">=</span> <span class="p">{</span><span class="mi">1</span><span class="p">:{</span><span class="s">&#39;WT&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;COUNTS&#39;</span><span class="p">:</span><span class="mi">7</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span>
<span class="gp">... </span>     <span class="s">&#39;a&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;t&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;c&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;g&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;N&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">},</span> <span class="mi">2</span><span class="p">:{</span><span class="s">&#39;WT&#39;</span><span class="p">:</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;COUNTS&#39;</span><span class="p">:</span><span class="mi">8</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">6</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;a&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;t&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;c&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;g&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;N&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteNTCounts</span><span class="p">(</span><span class="n">nt_counts</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">ReadNTCounts</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">==</span> <span class="n">nt_counts</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.ReadReadMismatches">
<tt class="descclassname">src.io.</tt><tt class="descname">ReadReadMismatches</tt><big>(</big><em>f</em><big>)</big><a class="headerlink" href="#src.io.ReadReadMismatches" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads read mismatches written by <cite>WriteReadMismatches</cite>.</p>
<p><cite>f</cite> : should be a readable file-like object, opened to the position
where output was written by <cite>WriteReadMismatches</cite>.</p>
<p>This function returns a dictionary <cite>d</cite> containing information about
the distribution of read mismatches along the read from the data
written to <cite>f</cite> by <cite>WriteReadMismatches</cite>. The dictionary is keyed by read
positions (in 1, 2, ... numbering) and the values are the fraction
of mismatches at that position among all reads.</p>
<p>Here is an example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">rlengths</span> <span class="o">=</span> <span class="p">{</span><span class="mi">10</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">11</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">12</span><span class="p">:</span><span class="mi">2</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rmismatches</span> <span class="o">=</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="mi">8</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">11</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="mi">12</span><span class="p">:</span><span class="mi">1</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">cStringIO</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteReadMismatches</span><span class="p">(</span><span class="n">rlengths</span><span class="p">,</span> <span class="n">rmismatches</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span><span class="mf">0.285714</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span><span class="mf">0.142857</span><span class="p">,</span> <span class="mi">4</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span> <span class="mi">5</span><span class="p">:</span><span class="mf">0.142857</span><span class="p">,</span>
<span class="gp">... </span>     <span class="mi">6</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span> <span class="mi">7</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span> <span class="mi">8</span><span class="p">:</span><span class="mf">0.285714</span><span class="p">,</span> <span class="mi">9</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span> <span class="mi">10</span><span class="p">:</span><span class="mf">0.0</span><span class="p">,</span> <span class="mi">11</span><span class="p">:</span><span class="mf">0.2</span><span class="p">,</span> <span class="mi">12</span><span class="p">:</span><span class="mf">0.5</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">d</span> <span class="o">==</span> <span class="n">ReadReadMismatches</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.WriteAAsFromCodonCounts">
<tt class="descclassname">src.io.</tt><tt class="descname">WriteAAsFromCodonCounts</tt><big>(</big><em>infile</em>, <em>outfile</em><big>)</big><a class="headerlink" href="#src.io.WriteAAsFromCodonCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads in codon counts, writes out amino acid counts.</p>
<p><cite>infile</cite> : name of an existing file of the type created by 
<cite>WriteCodonCounts</cite>.</p>
<p><cite>outfile</cite> : name of the resulting output file, which is created.</p>
<p>Reads in codon counts from <cite>infile</cite>. For all definitively called
codons (all upper case nucleotides) writes the number of
amino acids of that identity. Also writes the total number
of definitively called codons at that position, and the wildtype
amino acid. The format is self-explanatory.</p>
</dd></dl>

<dl class="function">
<dt id="src.io.WriteAlignmentStatistics">
<tt class="descclassname">src.io.</tt><tt class="descname">WriteAlignmentStatistics</tt><big>(</big><em>s</em>, <em>f</em><big>)</big><a class="headerlink" href="#src.io.WriteAlignmentStatistics" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes summary of statistics from <cite>MakeAlignment</cite> attempt.</p>
<p>This function writes a summary of the attempted / successful read
alignment statistics to a writeable file-like object.</p>
<p><cite>s</cite> : a dictionary holding the statistics in key / value format, with
string keys and integer values. The specific keys are described
below.</p>
<p><cite>f</cite> : a writeable file-like object to which we write the summary.</p>
<p><cite>s</cite> should have the following keys. These constitute the data written
to <cite>f</cite>. Their meanings are explained in the example below:</p>
<blockquote>
<div><p>&#8216;nread&#8217;</p>
<p>&#8216;nfiltered&#8217;</p>
<p>&#8216;nlowq&#8217;</p>
<p>&#8216;nexcessn&#8217;</p>
<p>&#8216;nattempted&#8217;</p>
<p>&#8216;npaired&#8217;</p>
<p>&#8216;nalignedfullgene&#8217;</p>
<p>&#8216;nalignedgene&#8217;</p>
</div></blockquote>
<p>Here is example output, comment lines beginning with # introduce
each variable, which is then named and followed by its value.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;nread&#39;</span><span class="p">:</span><span class="mi">110</span><span class="p">,</span> <span class="s">&#39;nfiltered&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="s">&#39;nlowq&#39;</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span> <span class="s">&#39;nexcessn&#39;</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span> <span class="s">&#39;nattempted&#39;</span><span class="p">:</span><span class="mi">90</span><span class="p">,</span>
<span class="gp">... </span>   <span class="s">&#39;npaired&#39;</span><span class="p">:</span><span class="mi">65</span><span class="p">,</span> <span class="s">&#39;nalignedfullgene&#39;</span><span class="p">:</span><span class="mi">55</span><span class="p">,</span> <span class="s">&#39;nalignedgene&#39;</span><span class="p">:</span><span class="mi">50</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteAlignmentStatistics</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>
<span class="go"># nread -&gt; total number of read pairs read from the FASTQ files.</span>
<span class="go">nread 110</span>
<span class="go"># nfiltered -&gt; number of read pairs removed by the Illumina filter.</span>
<span class="go">nfiltered 10</span>
<span class="go"># nlowq -&gt; number of read pairs removed due to low mean Q-score</span>
<span class="go">nlowq 5</span>
<span class="go"># nexcessn -&gt; number of read pairs removed due to excess N/n nucleotides</span>
<span class="go">nexcessn 5</span>
<span class="go"># nattempted -&gt; number of read pairs we attempted to align</span>
<span class="go">nattempted 90</span>
<span class="go"># npaired -&gt; number of read pairs where R1 and R2 were successfully</span>
<span class="go"># overlapped</span>
<span class="go">npaired 65</span>
<span class="go"># nalignedfullgene -&gt; number of read pairs where the overlapped read</span>
<span class="go"># pairs were successfully aligned to fullgene (the template</span>
<span class="go"># amplicon).</span>
<span class="go">nalignedfullgene 55</span>
<span class="go"># nalignedgene -&gt; number of read pairs which overlapped with the</span>
<span class="go"># gene of interest (region of fullgene specified by generange).</span>
<span class="go">nalignedgene 50</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.WriteCodonCounts">
<tt class="descclassname">src.io.</tt><tt class="descname">WriteCodonCounts</tt><big>(</big><em>codon_counts</em>, <em>f</em><big>)</big><a class="headerlink" href="#src.io.WriteCodonCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes counts of codon identities.</p>
<p>This function takes a dictionary that specifies the number of
times each codon identity was observed at a position. This
is used to write a text output file.</p>
<p>This function operates exactly like <cite>WriteNTCounts</cite>, except
that it operates on codon rather than nucleotide counts,
so <cite>codon_counts</cite> should be keyed by codons rather than
nucleotides.</p>
<p>So the required keys for <cite>codon_counts[i]</cite> for each codon <cite>i</cite> (1, 2, ...
numbering) are:</p>
<blockquote>
<div><p>&#8216;WT&#8217; : value should be an upper-case string indicating the
wildtype identity, such as &#8216;ATG&#8217;.</p>
<p>&#8216;COUNTS&#8217; : integer total number of counts for this codon.</p>
<p>All codon codes in <cite>AllCodons()</cite>. For these codes, upper-case
letters (A, T, C, G) specify identities called by both reads.
Lower-case letter (a, t, c, g) specify identities called by
just one read, with the other read ambiguous or excluded at
the relevant position. N specifies identities where the 
two reads disagree, both are ambiguous, or both are excluded.
If any of the required keys are missing, raises a <cite>ValueError</cite>.</p>
</div></blockquote>
<p>So upper-case letters indicate positions called by both
reads, lower-case letters indicate positions called by
one read and the other ambiguous, and N indicates disagreement
between reads or both ambiguous.</p>
<p>The codons are listed in the order returned by <cite>AllCodons()</cite>.</p>
</dd></dl>

<dl class="function">
<dt id="src.io.WriteInsertLengths">
<tt class="descclassname">src.io.</tt><tt class="descname">WriteInsertLengths</tt><big>(</big><em>d</em>, <em>f</em><big>)</big><a class="headerlink" href="#src.io.WriteInsertLengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes distribution of paired read insert lengths.</p>
<p>This function writes a text summary of the distribution of different
insert lengths found between successfully overlapped paired-end
reads. The insert length is the size of the template fragment
between the two read adaptors.</p>
<p><cite>d</cite> : dictionary holding the distribution of insert lengths. It is
keyed by integers giving the insert lengths. For each length,
the value is an integer giving the number of times inserts of
this length are found. This method will raise a <cite>ValueError</cite> if <cite>d</cite>
is empty or records no statistics.</p>
<p><cite>f</cite> : writeable file-like object to which we write the summary.</p>
<p>Here is example output, comment lines beginning with # introduce
head each column.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="mi">20</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">21</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span> <span class="mi">22</span><span class="p">:</span><span class="mi">7</span><span class="p">,</span> <span class="mi">23</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">24</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span> <span class="mi">25</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">27</span><span class="p">:</span><span class="mi">1</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteInsertLengths</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>
<span class="go"># Distribution of insert lengths</span>
<span class="go"># Length Number Fraction</span>
<span class="go">20 3 0.0909</span>
<span class="go">21 5 0.1515</span>
<span class="go">22 7 0.2121</span>
<span class="go">23 10 0.3030</span>
<span class="go">24 5 0.1515</span>
<span class="go">25 2 0.0606</span>
<span class="go">26 0 0.0000</span>
<span class="go">27 1 0.0303</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.WriteNTCounts">
<tt class="descclassname">src.io.</tt><tt class="descname">WriteNTCounts</tt><big>(</big><em>nt_counts</em>, <em>f</em><big>)</big><a class="headerlink" href="#src.io.WriteNTCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes counts of nucleotide identities.</p>
<p>This function takes a dictionary that specifies the number of
times each nucleotide identity was observed at a position. This
is used to write a text output file.</p>
<p><cite>nt_counts</cite> : a dictionary keyed by nucleotide number. These integer
keys should range from 1 to <cite>max(nt_counts.keys())</cite> inclusive, and
there must be a key for every number within this range. For key
<cite>nti</cite>, <cite>nt_counts[nti]</cite> is another dictionary specifying the counts
for this specific nucleotide. This nucleotide dictionary must
have the following string keys:</p>
<blockquote>
<div><p>&#8216;WT&#8217; : value should be a letter indicating wildtype nucleotide
identity (such as &#8216;A&#8217;, &#8216;T&#8217;, &#8216;C&#8217;, or &#8216;G&#8217;)</p>
<p>&#8216;COUNTS&#8217; -&gt; integer total number of counts for this nucleotide</p>
<p>All nucleotide codes in <cite>AllNTs()</cite>. For these codes, upper-case
letters (A, T, C, G) specify identities called by both reads.
Lower-case letter (a, t, c, g) specify identities called by
just one read, with the other read ambiguous or excluded at
the relevant position. N specifies identities where the 
two reads disagree, both are ambiguous, or both are excluded.
If any of the required keys are missing, raises a <cite>ValueError</cite>.</p>
</div></blockquote>
<p><cite>f</cite> : writable file like object to which we write the counts.</p>
<p>Here is example output, comment lines beginning with # introduce
each variable, which is then named and followed by its value. The
order of the nucleotide entries matches the order returned by <cite>AllNTs()</cite>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">nt_counts</span> <span class="o">=</span> <span class="p">{</span><span class="mi">1</span><span class="p">:{</span><span class="s">&#39;WT&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;COUNTS&#39;</span><span class="p">:</span><span class="mi">7</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span>
<span class="gp">... </span>     <span class="s">&#39;a&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;t&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;c&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;g&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;N&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">},</span> <span class="mi">2</span><span class="p">:{</span><span class="s">&#39;WT&#39;</span><span class="p">:</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;COUNTS&#39;</span><span class="p">:</span><span class="mi">8</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mi">6</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;a&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;t&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;c&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;g&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;N&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteNTCounts</span><span class="p">(</span><span class="n">nt_counts</span><span class="p">,</span> <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>
<span class="go"># nucleotide counts</span>
<span class="go"># upper case indicates nucleotide called by both reads</span>
<span class="go"># lower case indicates nucleotide called by one read with other ambiguous</span>
<span class="go"># N indicates disagreement between reads, or both ambiguous</span>
<span class="go"># POSITION WT COUNTS A T C G a t c g N</span>
<span class="go">1 A 7 3 1 0 0 1 1 0 0 1</span>
<span class="go">2 C 8 0 0 6 0 1 0 1 0 0</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="src.io.WriteReadMismatches">
<tt class="descclassname">src.io.</tt><tt class="descname">WriteReadMismatches</tt><big>(</big><em>rlengths</em>, <em>rmismatches</em>, <em>f</em><big>)</big><a class="headerlink" href="#src.io.WriteReadMismatches" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes distribution of mismatches along the read length.</p>
<p>This function writes a text summary of the distribution of
mismatches with a target gene as a function of position along the
read.</p>
<p><cite>rlengths</cite> : a dictionary keyed by the length of the reads that are
being analyzed for mismatches, and with values being the number
of reads of this length. For example, <tt class="docutils literal"><span class="pre">{25:2,</span> <span class="pre">26:3}</span></tt> indicates
that we have 2 reads of length 25 and 3 reads of length 26. All
reads are assumed to begin at the first position and extend up
to the indicated length.</p>
<p><cite>rmismatches</cite> : a dictionary giving the position of mismatches along
the read. The key is the read position in 1, 2, ... numbering.
The values are the number of mismatches found at that position
in the read.</p>
<p><cite>f</cite> : a writeable file-like object to which we write the summary.</p>
<p>Here is example output, comment lines beginning with # introduce
head each column:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">rlengths</span> <span class="o">=</span> <span class="p">{</span><span class="mi">10</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">11</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">12</span><span class="p">:</span><span class="mi">2</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rmismatches</span> <span class="o">=</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="mi">8</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">11</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="mi">12</span><span class="p">:</span><span class="mi">1</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">WriteReadMismatches</span><span class="p">(</span><span class="n">rlengths</span><span class="p">,</span> <span class="n">rmismatches</span><span class="p">,</span> <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>
<span class="go"># Distribution of mismatches along the read length.</span>
<span class="go"># Read positions numbered 1, 2, ... from read start.</span>
<span class="go"># ReadPosition NReads NMismatches FracMismatches</span>
<span class="go">1 7 2 0.285714</span>
<span class="go">2 7 0 0.000000</span>
<span class="go">3 7 1 0.142857</span>
<span class="go">4 7 0 0.000000</span>
<span class="go">5 7 1 0.142857</span>
<span class="go">6 7 0 0.000000</span>
<span class="go">7 7 0 0.000000</span>
<span class="go">8 7 2 0.285714</span>
<span class="go">9 7 0 0.000000</span>
<span class="go">10 7 0 0.000000</span>
<span class="go">11 5 1 0.200000</span>
<span class="go">12 2 1 0.500000</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><tt class="docutils literal"><span class="pre">io</span></tt> Module</a><ul>
<li><a class="reference internal" href="#list-of-functions">List of functions</a></li>
<li><a class="reference internal" href="#documentation-for-individual-functions">Documentation for individual functions</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="sequtils_module.html"
                        title="previous chapter"><tt class="docutils literal docutils literal"><span class="pre">sequtils</span></tt> Module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="stats_module.html"
                        title="next chapter"><tt class="docutils literal"><span class="pre">stats</span></tt> Module</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/io_module.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="stats_module.html" title="stats Module"
             >next</a> |</li>
        <li class="right" >
          <a href="sequtils_module.html" title="sequtils Module"
             >previous</a> |</li>
        <li><a href="index.html">mapmuts 1.1 documentation</a> &raquo;</li>
          <li><a href="pythonapi.html" >Python modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2013, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>