<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>mapmuts_makealignments.py &mdash; mapmuts 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="mapmuts 1.0 documentation" href="index.html" />
    <link rel="up" title="Scripts" href="scripts.html" />
    <link rel="next" title="mapmuts_alignmentsummaryplot.py" href="mapmuts_alignmentsummaryplot.html" />
    <link rel="prev" title="Scripts" href="scripts.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="mapmuts_alignmentsummaryplot.html" title="mapmuts_alignmentsummaryplot.py"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="scripts.html" title="Scripts"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">mapmuts 1.0 documentation</a> &raquo;</li>
          <li><a href="scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="mapmuts-makealignments-py">
<span id="id1"></span><h1>mapmuts_makealignments.py<a class="headerlink" href="#mapmuts-makealignments-py" title="Permalink to this headline">¶</a></h1>
<p>Script to align overlapping paired-end Illumina reads to a target gene.</p>
<p>This is one of the core scripts of the <a class="reference external" href="https://github.com/jbloom/mapmuts">mapmuts</a> package.</p>
<p>To run this script from the prompt, first create a text infile of the
format described below. Then simply type <a class="reference internal" href="#mapmuts-makealignments-py"><em>mapmuts_makealignments.py</em></a>
followed by the infile name. For example, if the name is <tt class="docutils literal"><span class="pre">infile.txt</span></tt>,
type:</p>
<div class="highlight-python"><div class="highlight"><pre>mapmuts_makealignments.py infile.txt
</pre></div>
</div>
<p>This script aligns overlapping paired-end sequencing reads to a target
gene sequence. It has been tested on the alignment to a roughly 1.5 kb
gene of 50 nt paired-end Illumina
reads in FASTQ files generated by the Casava 1.8 pipeline, although it should
work more generally.</p>
<p>The result of running this script is a variety of output
files containing the actual alignments and summary information about the
alignment process. If <a class="reference external" href="http://wiki.scipy.org/PyLab">pylab</a> / <a class="reference external" href="http://matplotlib.org/">matplotlib</a> are available, summary plots
are also generated. If <a class="reference external" href="http://www.tug.org/applications/pdftex/">pdflatex</a> is available, these plots are merged
into a summary PDF document as well.</p>
<p>This script runs according to an input text file, the format of which is
described below. This text file provides options for specifying the
stringency and parameters of the alignments.</p>
<div class="section" id="the-alignment-algorithm">
<h2>The alignment algorithm<a class="headerlink" href="#the-alignment-algorithm" title="Permalink to this headline">¶</a></h2>
<p>Some important notes about how the alignments are performed:</p>
<ul class="simple">
<li>The alignments performed by this script are not actually true global
alignments, but rather direct searches for substrings allowing
some mismatches. This is efficient for short reads that are being
aligned to relatively short genes (a few kb). It will become
increasingly less computationally efficient for longer reads or
target genes. In addition, these alignment methods do not consider
the possibility of gapped alignments. This tends not to be a problem
for Illumina reads, where mismatches are the main type of sequencing
error. It could be a problem for other types of sequencing data
where indel errors are common.</li>
<li>N / n nucleotides do not count as mismatches, regardless of what they
are aligned with. Instead, there is simply a limit on how many of
these can occur in a read as specified by maxn described below.</li>
<li>The alignments are case-sensitive. In general, you will want to
guarantee that the sequences are upper case. See the
<cite>upcase</cite> option described in the description of the input file below.</li>
</ul>
<p>The general algorithm is as follows:</p>
<ol class="arabic simple">
<li>Paired Illumina reads (R1 and R2 reads) are read from a FASTQ
file. If the <cite>applyfilter</cite> option is <cite>True</cite>, any reads that fail the
Illumina chastity filter are discarded. In addition, any reads
that fail the Q-score cutoff specified by <cite>minq</cite> are discarded.</li>
<li>The R1 and R2 reads are aligned to each other assuming that reads
have the terminal adaptor sequences specified by <cite>a1</cite> and <cite>a2</cite>. The
alignment must meet the standards specified by <cite>maxn</cite>, <cite>minoverlap</cite>,
<cite>maxrm</cite>, <cite>maxa1m</cite>, and <cite>maxa2m</cite>. If the paired reads cannot be aligned
to each other, they are discarded.</li>
<li>The overlapped region of the R1 and R2 reads is aligned to the
full target sequence given by fullgene. If the region cannot
be aligned with &lt;= <cite>maxgenem</cite> mismatches, the reads are discarded.</li>
<li>The portion of the overlap to the gene range of interest
(specified by <cite>generange</cite>) is written to an output file, assuming
that some of the overlap is to the gene range of interest.</li>
<li>Other output text files are generated with statistics from the
alignment process.</li>
<li>If <a class="reference external" href="http://wiki.scipy.org/PyLab">pylab</a> / <a class="reference external" href="http://matplotlib.org/">matplotlib</a> are available, summary plots are made.</li>
<li>If <a class="reference external" href="http://www.tug.org/applications/pdftex/">pdflatex</a> is available, an overall summary of the alignment
statistics is made.</li>
</ol>
</div>
<div class="section" id="format-of-the-input-file">
<h2>Format of the input file<a class="headerlink" href="#format-of-the-input-file" title="Permalink to this headline">¶</a></h2>
<p>The input file specifies <cite>key</cite> / <cite>value</cite> pairs on individual lines in the format:</p>
<div class="highlight-python"><div class="highlight"><pre>key1 value1
key2 value2
</pre></div>
</div>
<p>Blank lines or lines
that begin with # are ignored (i.e. as comment lines). The input file should
contain the following keys:</p>
<ul>
<li><p class="first">The keys <cite>r1files</cite> and <cite>r2files</cite> specify the input FASTQ files, which can
potentially be gzipped. After the key name should
be a list of the name (including directory path if not in the
current directory) of one or more FASTQ files. The <cite>r1files</cite> should
contain the first (R1) reads of a paired-end set, and the <cite>r2files</cite>
should contain the second (R2) reads of a paired-end set. Note that
the entries in these files must be in the same order, and each must
contain the same number of reads, so that when reads are read
sequentially from <cite>r1files</cite> they pair with the reads read sequentially
from <cite>r2files</cite>. In addition to specifically specifying the file names, you can also list general path information such as <tt class="docutils literal"><span class="pre">./my_directory/*_R1*_.fastq.gz</span></tt> and the <tt class="docutils literal"><span class="pre">*</span></tt> character will be expanded to mean any value as in the standard Linux file naming. In this case, the lists of files matching the specification for <em>r1files</em> and <em>r2files</em> must be the same, and when sorted in alphabetical order the files must pair with each other when the reads are read sequentially. If you are using this sort of general name specification then you can only list one name each for <em>r1files</em> and <em>r2files</em>.</p>
</li>
<li><p class="first"><cite>gzipped</cite> is a Boolean switch specifying whether the FASTQ files
specified by <cite>r1files</cite> and <cite>r2files</cite> are gzipped. If it is set to <cite>True</cite>,
these files should be gzipped. If it is set to <cite>False</cite>, they should not
be gzipped.</p>
</li>
<li><p class="first"><cite>applyfilter</cite> is a Boolean switch specifying whether we discard read
pairs in which either the R1 or R2 read failed the Illumina chastity
filter. If <cite>True</cite>, all reads that failed the filter are discarded. If
<cite>False</cite>, they are not. In general, you probably want this option to be
<cite>True</cite> unless you have a good reason for disregarding the Illumina
chastity filter.</p>
</li>
<li><p class="first"><cite>minq</cite> specifies the minimum acceptable average Q-score (with the
average taken over all positions in the read). Any read pair in
which either read has an average Q-score &lt; <cite>minq</cite> is discarded. While
setting very high values of <cite>minq</cite> might be supposed to give more
accurate reads, in practice we have found that high values of <cite>minq</cite> do
not seem to improve accuracy much, apparently because problematic
reads tend to be removed in a more sensitive and selective manner by
setting reasonably stringent values for the other alignment parameters
below. Therefore, a modest value such as 20 or 25 is probably fine.</p>
</li>
<li><p class="first"><cite>fullgenefile</cite> specifies the name of the file containing the full target
sequence to which the reads will be aligned. In general, this is the
PCR amplicon that was fragmented to prepare the reads. Typically, this
full amplicon sequence will be longer than the actual gene of
interest, since the PCR creation of the amplicon often adds extra
terminal sequences. For successful alignment, it is important that
this file specifies the full amplicon sequence, not just the gene
region of interest. That region is specified by the argument
<cite>generange</cite>. <cite>fullgenefile</cite> should be a FASTA file containing a single
entry.</p>
</li>
<li><p class="first"><cite>generange</cite> specifies the portion of fullgene that contains the gene of
interest. The gene sequence is numbered sequentially 1, 2, 3, ...
The <cite>generange</cite> key should be followed by two numbers. The first number
gives the number of the first nucleotide of <cite>fullgene</cite> that is part of
the gene of interest, and the last number gives the number of last
nucleotide that is part of the gene of interest. The terminal stop
codon should not be included.</p>
</li>
<li><p class="first"><cite>a1file</cite> specifies the name of the file containing the adaptor sequence
that is found at the 3&#8217; end of any R1 reads that read past the insert
and into the adaptor. <cite>a2file</cite> specifies the same for R2 reads. Both of
these keys should specify FASTA files containing a single entry giving
the adaptor.</p>
</li>
<li><p class="first"><cite>maxn</cite> specifies the maximum number of N or n nucleotides that are
allowed in either read of a read pair. If either read has &gt; <cite>maxn</cite> N or
n nucleotides, the read pair is discarded.</p>
</li>
<li><p class="first"><cite>minoverlap</cite> specifies the minimum length of the overlap between R1 and
R2, representing their joint coverage of the target sequence. The
overlap between the two reads must be &gt;= this number.</p>
</li>
<li><p class="first"><cite>maxrm</cite> specifies the maximum number of mismatches allowed in the overlap
of R1 and R2. The number of mismatches in the overlapped region of the
two reads must be &lt;= this number.</p>
</li>
<li><p class="first"><cite>maxa1m</cite> specifies the maximum number of mismatches allowed in the overlap
of R1 with adaptor <cite>a1</cite>. The number of mismatches must be &lt;= this
number. <cite>maxa2m</cite> specifies the same for R2 and its adaptor.</p>
</li>
<li><p class="first"><cite>maxgenem</cite> specifies the maximum number of mismatches that is allowed for
either read individually when it is aligned with <cite>fullgene</cite> after
removing the read adaptor sequences (<cite>a1</cite> or <cite>a2</cite>). The read pair is
discarded if either read has &gt; this many mismatches when aligned with
the full target gene.</p>
</li>
<li><p class="first"><cite>upcase</cite> specifies how we handle possible upper / lower case differences
in nucleotide codes. This is important since the alignment procedures
are case-sensitive (i.e. a is different than A). If <cite>upcase</cite> is set to
<cite>True</cite>, then all sequences are converted to upper case before alignment.
This is the safest option, but is also the slowest since it requires
so many case conversions. Another option is <cite>test</cite>, which means that the
target and adaptor sequences (<cite>fullgene</cite>, <cite>a1</cite>, <cite>a2</cite>) are converted to upper
case, and the FASTQ files are tested to make sure the first entry is
upper case (as is typically the case for Illumina read). If the first
entry is upper case, no case conversion is done on the reads, which
will be faster. If the first entry is not upper case, then an
exception is raised, and you will probably have to switch this option
to <cite>True</cite>.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite> specifies the prefix appended to all output files created
by this script. These output files are detailed in the next section below.
Any existing files are overwritten.</p>
</li>
<li><p class="first"><cite>samplename</cite> specifies the name of the sample for this run of the
script. Name are allowed to contain spaces.</p>
</li>
<li><dl class="first docutils">
<dt><cite>write_unaligned</cite> specifies that we write all unaligned reads to the</dt>
<dd><p class="first last"><tt class="docutils literal"><span class="pre">*_unaligned.fasta.gz</span></tt> file. This is done if set to <cite>True</cite>, not
done if set to <cite>False</cite>. Each entry also contains a brief explanation of why it failed to align.</p>
</dd>
</dl>
</li>
</ul>
<p>Here is an example input file:</p>
<div class="highlight-python"><div class="highlight"><pre># Example input file for mapmuts_makealignments.py
maxa2m 1
generange 62 1555
write_unaligned True
r2files /home/jbloom/mapmuts/examples/2013Analysis_Influenza_NP_Aichi68/FASTQ_files/replicate_A/Sample_WT-1_DNA/*R2*.fastq.gz
maxa1m 1
a1file /home/jbloom/mapmuts/examples/2013Analysis_Influenza_NP_Aichi68/R1_trim3.fasta
gzipped True
minq 25
minoverlap 30
maxgenem 6
r1files /home/jbloom/mapmuts/examples/2013Analysis_Influenza_NP_Aichi68/FASTQ_files/replicate_A/Sample_WT-1_DNA/*R1*.fastq.gz
a2file /home/jbloom/mapmuts/examples/2013Analysis_Influenza_NP_Aichi68/R2_trim3.fasta
maxrm 1
outfileprefix replicate_A_WT-1_DNA
fullgenefile /home/jbloom/mapmuts/examples/2013Analysis_Influenza_NP_Aichi68/Aichi68-NP_amplicon.fasta
maxn 2
upcase test
applyfilter True
samplename replicate_A, WT-1, DNA
</pre></div>
</div>
</div>
<div class="section" id="output-files">
<h2>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>Running this script produces a series of output files. The names of these files all include the prefix specified by <cite>outfileprefix</cite> in the input file, and then the indicated suffix.
Any existing files are overwritten.</p>
<ul>
<li><p class="first"><cite>outfileprefix</cite>_makealignments_log.txt : a text file that tracks the output of this script.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_alignments.txt.gz : A gzipped text file containing the read pair to
gene alignments for all read pairs that pass the various filters
and cutoffs. Each alignment consists of five lines.
These lines give:</p>
<blockquote>
<div><ol class="arabic simple">
<li>the read name,</li>
<li>the alignment region of the gene with indices,</li>
<li>the alignment region of R1 with indices,</li>
<li>the alignment region of R2 with indices,</li>
<li>a blank line.</li>
</ol>
</div></blockquote>
<p>The indices indicate the first and last nucleotide shown in
1, 2, ... numbering. For the read, a . character is used to
indicate identities with the gene at that position. The alignment
region is shown only for regions of gene (the portion of <cite>fullgene</cite>
specified by <cite>generange</cite>) covered by both reads. For example:</p>
<div class="highlight-python"><div class="highlight"><pre>@DH1DQQN1:241:C1433ACXX:2:1101:1623:2365
14 ATAGATACTAGAT 26
1 .....C.....T. 13
13 .....C.......  1
</pre></div>
</div>
</li>
<li><p class="first"><cite>outfileprefix</cite>_alignmentstatistics.txt : A text file giving a summary of the
alignment statistics (i.e. how many reads were successfully
aligned and how many were filtered by the various cutoffs) in what
should be a self-explanatory format.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_insertlengths.txt : A text file giving the distribution of insert
lengths for all R1-R2 read pairs that align with each other
according to the given specifications. The shortest possible value
of the insert length is <cite>minoverlap</cite>, while the longest possible
value is the sume of the two read lengths minus <cite>minoverlap</cite>.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_R1mismatches.txt : A text file giving the fraction and number of
mismatches between the R1 read and the <cite>fullgene</cite> template for all
positions in the read that align with <cite>fullgene</cite> after passing the
various filtration criteria. Read positions are numbered 1, 2, ...</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_R2mismatches.txt : Like <cite>_R1mismatches.txt</cite>, but for R2 reads.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_alignmentstatistics.pdf : Created if <a class="reference external" href="http://wiki.scipy.org/PyLab">pylab</a> / <a class="reference external" href="http://matplotlib.org/">matplotlib</a> is available, a graphical
summary of data in <cite>_alignmentstatistics.txt</cite>.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_insertlengths.pdf : Created if <a class="reference external" href="http://wiki.scipy.org/PyLab">pylab</a> / <a class="reference external" href="http://matplotlib.org/">matplotlib</a> is available, a graphical
summary of data in <cite>_insertlengths.txt</cite>.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_R1mismatches.pdf : Created if <a class="reference external" href="http://wiki.scipy.org/PyLab">pylab</a> / <a class="reference external" href="http://matplotlib.org/">matplotlib</a> is available, a graphical
summary of data in <cite>_R1mismatches.txt</cite>.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_R2mismatches.pdf : Created if <a class="reference external" href="http://wiki.scipy.org/PyLab">pylab</a> / <a class="reference external" href="http://matplotlib.org/">matplotlib</a> is available, a graphical
summary of data in <cite>_R2mismatches.txt</cite>.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_makealignments_summary.pdf : created if <a class="reference external" href="http://wiki.scipy.org/PyLab">pylab</a> / <a class="reference external" href="http://matplotlib.org/">matplotlib</a> and <a class="reference external" href="http://www.tug.org/applications/pdftex/">pdflatex</a> are available, an
overall graphical summary of the alignment statistics, insert
lengths, mismatch positions, and filtering criteria.</p>
</li>
<li><p class="first"><cite>outfileprefix</cite>_unaligned.fasta.gz : created if <cite>write_unaligned</cite> is <cite>True</cite>, a gzipped
FASTA file with all unaligned reads. Each entry also contains a brief explanation of why it failed to align.</p>
</li>
</ul>
</div>
<div class="section" id="example-of-summary-plot">
<h2>Example of summary plot<a class="headerlink" href="#example-of-summary-plot" title="Permalink to this headline">¶</a></h2>
<p>Here is an example of the <cite>outfileprefix</cite>_makealignments_summary.pdf plot.</p>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/example_makealignments_summary.jpg"><img alt="example_makealignments_summary.jpg" src="_images/example_makealignments_summary.jpg" style="width: 60%;" /></a>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">mapmuts_makealignments.py</a><ul>
<li><a class="reference internal" href="#the-alignment-algorithm">The alignment algorithm</a></li>
<li><a class="reference internal" href="#format-of-the-input-file">Format of the input file</a></li>
<li><a class="reference internal" href="#output-files">Output files</a></li>
<li><a class="reference internal" href="#example-of-summary-plot">Example of summary plot</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="scripts.html"
                        title="previous chapter">Scripts</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="mapmuts_alignmentsummaryplot.html"
                        title="next chapter">mapmuts_alignmentsummaryplot.py</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/mapmuts_makealignments.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="mapmuts_alignmentsummaryplot.html" title="mapmuts_alignmentsummaryplot.py"
             >next</a> |</li>
        <li class="right" >
          <a href="scripts.html" title="Scripts"
             >previous</a> |</li>
        <li><a href="index.html">mapmuts 1.0 documentation</a> &raquo;</li>
          <li><a href="scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2013, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>