<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>plot Module &mdash; mapmuts 1.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="mapmuts 1.1 documentation" href="index.html" />
    <link rel="up" title="Python modules" href="pythonapi.html" />
    <link rel="next" title="latex Module" href="latex_module.html" />
    <link rel="prev" title="simulate Module" href="simulate_module.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="latex_module.html" title="latex Module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="simulate_module.html" title="simulate Module"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">mapmuts 1.1 documentation</a> &raquo;</li>
          <li><a href="pythonapi.html" accesskey="U">Python modules</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-plot">
<span id="plot-module"></span><h1><a class="reference internal" href="#module-plot" title="plot"><tt class="xref py py-mod docutils literal"><span class="pre">plot</span></tt></a> Module<a class="headerlink" href="#module-plot" title="Permalink to this headline">¶</a></h1>
<p>Module for performing plotting for <tt class="docutils literal"><span class="pre">mapmuts</span></tt> package.</p>
<p>This module uses <tt class="docutils literal"><span class="pre">pylab</span></tt> and <tt class="docutils literal"><span class="pre">matplotlib</span></tt> to make plots. These plots will
fail if <tt class="docutils literal"><span class="pre">pylab</span></tt> and <tt class="docutils literal"><span class="pre">matplotlib</span></tt> are not available for importation. Before
running any function in this module, you can run the <cite>PylabAvailable</cite>
function to determine if <tt class="docutils literal"><span class="pre">pylab</span></tt> and <tt class="docutils literal"><span class="pre">matplotlib</span></tt> are available. Otherwise,
calling any other function will raise an Exception if thise modules are
not available. The <tt class="docutils literal"><span class="pre">pdf</span></tt> backend is used for <tt class="docutils literal"><span class="pre">matplotlib</span></tt> / <tt class="docutils literal"><span class="pre">pylab</span></tt>. This means
that plots must be created as PDF files.</p>
<p>The <em>PlotCorrelation</em> function can use <tt class="docutils literal"><span class="pre">scipy</span></tt> to calculate correlation
coefficients if this package is available.</p>
<p>Written by Jesse Bloom.</p>
<div class="section" id="list-of-functions">
<h2>List of functions<a class="headerlink" href="#list-of-functions" title="Permalink to this headline">¶</a></h2>
<p><cite>PylabAvailable</cite> : tests whether <tt class="docutils literal"><span class="pre">pylab</span></tt> and <tt class="docutils literal"><span class="pre">matplotlib</span></tt> can be imported.</p>
<p><cite>PlotAlignmentStatistics</cite></p>
<p><cite>PlotInsertLengths</cite></p>
<p><cite>PlotReadMismatches</cite></p>
<p><cite>PlotCodonDepth</cite></p>
<p><cite>PlotCodonDist</cite></p>
<p><cite>PlotMutFracs</cite></p>
<p><cite>PlotPairedMutFracs</cite></p>
<p><cite>PlotNTMutFracs</cite></p>
<p><cite>PlotAAFracs</cite></p>
<p><cite>PlotMutCountFracs</cite> : plots fraction mutations with at least a certain # of counts.</p>
<p><cite>PlotEnrichmentRatios</cite> : plots enrichment ratios on log scale</p>
<p><cite>PlotEquilibriumFreqs</cite> : plots equilibrium frequencies of amino acids</p>
<p><cite>EquilibriumFreqsHeatMap</cite> : plots heat map of equilibrium amino-acid frequencies.</p>
<p><cite>EquilibriumFreqsLogo</cite> : sequence logo of amino-acid frequencies - DOES NOT WORK!</p>
<p><cite>LogoOverlay</cite> : color maps of RSA and SS to be overlayed on weblogo sequence logo</p>
<p><cite>PlotTraces</cite> : plots traces of posterior probabilities</p>
<p><cite>PlotCorrelation</cite> : plots correlation between two variables.</p>
<p><cite>PlotLinearDensity</cite> : plots one or more variables along primary sequence.</p>
<p><cite>Base10Formatter</cite></p>
<p><cite>SplitLabel</cite></p>
<p><cite>KyteDoolittleColorMapping</cite> : maps Kyte-Doolittle scale to colors.</p>
</div>
<div class="section" id="documentation-for-functions">
<h2>Documentation for functions<a class="headerlink" href="#documentation-for-functions" title="Permalink to this headline">¶</a></h2>
<p>Documentation for individual functions is provided in their definitions below.</p>
<dl class="function">
<dt id="plot.Base10Formatter">
<tt class="descclassname">plot.</tt><tt class="descname">Base10Formatter</tt><big>(</big><em>number</em>, <em>exp_cutoff</em>, <em>exp_decimal_digits</em>, <em>decimal_digits</em><big>)</big><a class="headerlink" href="#plot.Base10Formatter" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a number into Latex formatting with scientific notation.</p>
<p>Takes a number and converts it to a string that can be shown
in LaTex using math mode. It is converted to scientific notation
if the criteria specified by <cite>exp_cutoff</cite> is met.</p>
<p><cite>number</cite> : the number to be formatted, should be a float or integer.
Currently only works for <cite>number &gt;= 0</cite>.</p>
<p><cite>exp_cutoff</cite> : convert to scientific notation if <tt class="docutils literal"><span class="pre">abs(math.log10(number))</span> 
<span class="pre">&gt;=</span> <span class="pre">exp_cutoff</span></tt>.</p>
<p><cite>exp_decimal_digits</cite> : show this many digits after the decimal if number
is converted to scientific notation.</p>
<p><cite>decimal_digits</cite> : show this many digits after the decimal if number
is NOT converted to scientific notation.</p>
<p>The returned value is a LaTex formatted string. If the number is zero, the
returned string is simply &#8216;0&#8217;.</p>
<p>EXAMPLES:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mi">103</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;103.0&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">103.0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;1.0 \\times 10^{2}&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">103.0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;1.03 \\times 10^{2}&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">2892.3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> 
<span class="go">&#39;2.9 \\times 10^{3}&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> 
<span class="go">&#39;0&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="mf">0.012</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;1.2 \\times 10^{-2}&#39;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Base10Formatter</span><span class="p">(</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">number must be &gt;= 0</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="plot.EquilibriumFreqsHeatMap">
<tt class="descclassname">plot.</tt><tt class="descname">EquilibriumFreqsHeatMap</tt><big>(</big><em>sites</em>, <em>pi_d</em>, <em>aa_ordering</em>, <em>plotfile</em>, <em>otherprops=</em>, <span class="optional">[</span><span class="optional">]</span><em>interpolation='none'</em><big>)</big><a class="headerlink" href="#plot.EquilibriumFreqsHeatMap" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs a heat map show amino acid equilibrium frequencies.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt>. It will raise an exception if
<em>PylabAvailable()</em> is not <em>True</em>.</p>
<p>The heat map has 20 rows corresponding to each of the amino
acids, and as many columns as there are entries in <em>sites</em>
(one for each site in the protein). The heat map is colored
according to the equilibrium frequency of that site for
that amino acid, as specified by <em>pi_d</em>.</p>
<p>The heat map is generated using the <em>imshow</em> function
of <tt class="docutils literal"><span class="pre">pylab</span></tt>.</p>
<p>CALLING VARIABLES:</p>
<ul>
<li><p class="first"><em>sites</em> is a list of integers giving all of the sites that
are being included in the heat map. They must be consecutive,
so that <em>sites = [i for i in range(sites[0], sites[-1] + 1)]</em>.</p>
</li>
<li><p class="first"><em>pi_d</em> is a dictionary that has a key for every integer in
<em>sites</em>. The value of <em>pi_d[isite]</em> is itself a dictionary,
which has keys &#8216;PI_A&#8217;, &#8216;PI_C&#8217;, &#8216;PI_D&#8217;, etc for all 20
one-letter upper-case amino acid codes. The values
for these keys are the equilibrium frequencies of that
amino acid at that site. So <em>pi_d[isite][&#8216;PI_M&#8217;]</em> is
the equilibrium frequency for methionine at site <em>isite</em>.</p>
</li>
<li><p class="first"><em>aa_ordering</em> is a string specifying how the amino acids
are ordered in the rows from top to bottom. Typically you
might want them to roughly be ordered by physical property
for a more appealing layout. <em>aa_ordering</em> is a string that
can be passed to <em>mapmuts.sequtils.OrderedAminoAcids</em> to
get an ordered list of the amino acids. For example, valid 
values are &#8216;icelogo&#8217; and &#8216;hydrophobicity&#8217;.</p>
</li>
<li><p class="first"><em>plotfile</em> is a string giving the name of the PDF file containing
the heat map plot. It must end in the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
</li>
<li><p class="first"><em>otherprops</em> is an optional argument that can be used to specify
additional bars that are plotted above the main heat map
of the site preferences. By default, <em>otherprops</em> is an empty
list, which means that no additional bars are plotted. If you
want to plot additonal bars, then add tuple entries to the list.
The format of these tuples differs depending on whether the
data is scalar (such as RSA or site entropy) or categorical
(such as secondary structure):</p>
<blockquote>
<div><ul class="simple">
<li>Scalar data: each tuple should be <em>(propname, prop_d)</em> where <em>propname</em> 
is a string giving the name of the property and <em>prop_d</em> is a dictionary
keyed by the site number and with <em>prop_d[isite]</em> being the numerical
value of property <em>propname</em> for site <em>isite</em> for <em>isite</em> in <em>sites</em>.</li>
<li>Categorical data: each tuple should be <em>(propname, prop_d, categories)</em>
where <em>propname</em> is a string giving the name of the property,
<em>prop_d</em> is a dictionary keyed by the site number and with <em>prop_d[isite]</em>
being the value of property <em>propname</em> for site <em>isite</em> for <em>isite</em>
in <em>sites, and with *categories</em> being a listing of all of the categories
of values found in <em>prop_d</em>.</li>
</ul>
</div></blockquote>
<p>Unlike for <em>pi_d</em>, it is not required that <em>prop_d</em> have a key for 
all values of <em>isite</em>. Sites for which no value is specified
are shown in white.</p>
<p>Currently you are only allowed to specify a maximum of three
entries in <em>otherprops</em>.</p>
</li>
<li><p class="first"><em>interpolation</em> specifies how the colors are interpolated between
points on the heat map. You can pass any value that is acceptable
for the parameter of the same name in the <tt class="docutils literal"><span class="pre">pylab</span></tt> <em>imshow</em>
function. The default value is &#8216;none&#8217; which corresponds to
no interpolation.</p>
</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="plot.EquilibriumFreqsLogo">
<tt class="descclassname">plot.</tt><tt class="descname">EquilibriumFreqsLogo</tt><big>(</big><em>sites</em>, <em>pi_d</em>, <em>aa_ordering</em>, <em>plotfile</em>, <em>otherprops=</em>, <span class="optional">[</span><span class="optional">]</span><em>plotwidth=6.25</em>, <em>logoheight=0.7</em>, <em>nperline=71</em>, <em>tickevery=5</em><big>)</big><a class="headerlink" href="#plot.EquilibriumFreqsLogo" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs a sequence logo showing amino acid equilibrium frequencies.</p>
<p>THIS FUNCTION DOES NOT WORK. YOU WILL GET AN EXCEPTION IF YOU TRY TO RUN IT.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt>. It will raise an exception if
<em>PylabAvailable()</em> is not <em>True</em>.</p>
<p>CALLING VARIABLES:</p>
<ul>
<li><p class="first"><em>sites</em> is a list of integers giving all of the sites that
are being included in the heat map. They must be consecutive,
so that <em>sites = [i for i in range(sites[0], sites[-1] + 1)]</em>.</p>
</li>
<li><p class="first"><em>pi_d</em> is a dictionary that has a key for every integer in
<em>sites</em>. The value of <em>pi_d[isite]</em> is itself a dictionary,
which has keys &#8216;PI_A&#8217;, &#8216;PI_C&#8217;, &#8216;PI_D&#8217;, etc for all 20
one-letter upper-case amino acid codes. The values
for these keys are the equilibrium frequencies of that
amino acid at that site. So <em>pi_d[isite][&#8216;PI_M&#8217;]</em> is
the equilibrium frequency for methionine at site <em>isite</em>.</p>
</li>
<li><p class="first"><em>aa_ordering</em> is a string specifying how the amino acids
are ordered in in their color choices. Typically you
might want them to roughly be ordered by physical property
for a more appealing layout. <em>aa_ordering</em> is a string that
can be passed to <em>mapmuts.sequtils.OrderedAminoAcids</em> to
get an ordered list of the amino acids. For example, valid 
values are &#8216;icelogo&#8217; and &#8216;hydrophobicity&#8217;.</p>
</li>
<li><p class="first"><em>plotfile</em> is a string giving the name of the PDF file containing
the heat map plot. It must end in the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
</li>
<li><p class="first"><em>otherprops</em> is an optional argument that can be used to specify
additional bars that are plotted above the main heat map
of the site preferences. By default, <em>otherprops</em> is an empty
list, which means that no additional bars are plotted. If you
want to plot additonal bars, then add tuple entries to the list.
The format of these tuples differs depending on whether the
data is scalar (such as RSA or site entropy) or categorical
(such as secondary structure):</p>
<blockquote>
<div><ul class="simple">
<li>Scalar data: each tuple should be <em>(propname, prop_d)</em> where <em>propname</em> 
is a string giving the name of the property and <em>prop_d</em> is a dictionary
keyed by the site number and with <em>prop_d[isite]</em> being the numerical
value of property <em>propname</em> for site <em>isite</em> for <em>isite</em> in <em>sites</em>.</li>
<li>Categorical data: each tuple should be <em>(propname, prop_d, categories)</em>
where <em>propname</em> is a string giving the name of the property,
<em>prop_d</em> is a dictionary keyed by the site number and with <em>prop_d[isite]</em>
being the value of property <em>propname</em> for site <em>isite</em> for <em>isite</em>
in <em>sites, and with *categories</em> being a listing of all of the categories
of values found in <em>prop_d</em>.</li>
</ul>
</div></blockquote>
<p>Unlike for <em>pi_d</em>, it is not required that <em>prop_d</em> have a key for 
all values of <em>isite</em>. Sites for which no value is specified
are shown in white.</p>
<p>Currently you are only allowed to specify a maximum of three
entries in <em>otherprops</em>.</p>
</li>
<li><p class="first"><em>plotwidth</em> is the width of the plot in inches, is 6.25 by default.</p>
</li>
<li><p class="first"><em>logoheight</em> is the height of the logo in inches, 0.7 by default.</p>
</li>
<li><p class="first"><em>nperline</em> is the number of sites per line, 71 by default.</p>
</li>
<li><p class="first"><em>tickevery</em> specifies where the xtick label (residue number labels) go.
They are put at multiples of this number.</p>
</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="plot.KyteDoolittleColorMapping">
<tt class="descclassname">plot.</tt><tt class="descname">KyteDoolittleColorMapping</tt><big>(</big><em>maptype='jet'</em>, <em>reverse=True</em><big>)</big><a class="headerlink" href="#plot.KyteDoolittleColorMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps amino-acid hydrophobicities to colors.</p>
<p>Uses the Kyte-Doolittle hydrophobicity scale defined by 
<em>mapmuts.sequtils.KyteDoolittle</em>.</p>
<p>The returned variable is the 3-tuple <em>(cmap, mapping_d, mapper)</em>:</p>
<blockquote>
<div><ul class="simple">
<li><em>cmap</em> is a <tt class="docutils literal"><span class="pre">pylab</span></tt> <em>LinearSegmentedColorMap</em> object.</li>
<li><em>mapping_d</em> is a dictionary keyed by the one-letter amino-acid
codes. The values are the colors in CSS2 format (e.g. #FF0000
for red) for that amino acid. The value for a stop codon 
(denoted by a * character) is black (#000000).</li>
<li><em>mapper</em> is the actual <em>ScalarMappable</em> object.</li>
</ul>
</div></blockquote>
<p>The optional calling argument <em>maptype</em> should specify a valid
<tt class="docutils literal"><span class="pre">pylab</span></tt> color map. Is <em>jet</em> by default.</p>
<p>The optional calling argument <em>reverse</em> specifies that we set up the color
map so that the most hydrophobic residue comes first (in the Kyte-Doolittle
scale the most hydrophobic comes last as it has the largest value). This option
is <em>True</em> by default as it seems more intuitive to have charged residues red
and hydrophobic ones blue.</p>
</dd></dl>

<dl class="function">
<dt id="plot.LogoOverlay">
<tt class="descclassname">plot.</tt><tt class="descname">LogoOverlay</tt><big>(</big><em>sites</em>, <em>plotfile</em>, <em>rsa_d</em>, <em>ss_d</em>, <em>nperline</em>, <em>sitewidth</em>, <em>rmargin</em>, <em>logoheight</em>, <em>barheight</em>, <em>barspacing</em><big>)</big><a class="headerlink" href="#plot.LogoOverlay" title="Permalink to this definition">¶</a></dt>
<dd><p>Overlay for <em>weblogo.EquilibriumFreqsLogo</em> showing RSA and SS.</p>
<p>This function creates colored bars showing the relative
solvent accessibility (RSA) and secondary structure (SS) as 
well as legends for these and residue hydrophobicity that
can be overlayed onto the sequence logos created by
<em>weblogo.EquilibriumFreqsLogo</em>.</p>
<p>The trick of this function is to create the bars the right
size so they align when they overlay. This function works for
at least the cases on which it has been tested, but there
is some chance that the overlay will not work well for
other plots.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt>. It will raise an exception if
<em>PylabAvailable()</em> is not <em>True</em>.</p>
<p>CALLING VARIABLES:</p>
<ul class="simple">
<li><em>sites</em> is a list of integers giving all of the sites that
are being included in the logo. They must be consecutive,
so that <em>sites = [i for i in range(sites[0], sites[-1] + 1)]</em>.</li>
<li><em>plotfile</em> is a string giving the name of the PDF file containing
the overlay. It must end in the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</li>
<li><em>rsa_d</em> is a dictionary giving the RSA values. It is keyed by the
integer site numbers: <em>rsa_d[site]</em> gives a float with the RSA
for residue <em>site</em>. If <em>site</em> is not in RSA d, the value is
shown in white.</li>
<li><em>ss_d</em> is like <em>rsa_d</em> except that it gives the codes for the 
some discrete property which can be secondary structure or
something else. The values should be a string specifying
the property name, and there should be at leasat one
such property. For secondary structure, the 
values might be &#8216;helix&#8217;, &#8216;strand&#8217;, and &#8216;loop&#8217;. For
some other custom discrete property, these values
are whatever you specify.</li>
<li><em>nperline</em> is the number of sites per line.</li>
<li><em>sitewidth</em> is the width of each site in points.</li>
<li><em>rmargin</em> is the right margin in points.</li>
<li><em>logoheight</em> is the total height of each logo row in points.</li>
<li><em>barheight</em> is the total height of each bar in points.</li>
<li><em>barspacing</em> is the vertical spacing between bars in points.</li>
<li><em>cmap</em> is a pylab <em>LinearSegmentedColorMap</em> used for the bars.</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="plot.PlotAAFracs">
<tt class="descclassname">plot.</tt><tt class="descname">PlotAAFracs</tt><big>(</big><em>counts</em>, <em>names</em>, <em>icodon</em>, <em>plotfile</em>, <em>aas</em>, <em>title</em><big>)</big><a class="headerlink" href="#plot.PlotAAFracs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots frequency of different amino acid identities at a position.</p>
<p>For a given position in a protein, plots the fraction of all
definitively called codons (called by both reads) for a set
of samples.</p>
<p><cite>counts</cite> : a list of the codon_counts dictionaries of the type
returned by <cite>mapmuts.io.ReadCodonCounts</cite>.</p>
<p><cite>names</cite> : a list of strings giving the names of the samples
corresponding to each entry in counts. These names are shown
in the legend, augmented with numbers giving the total
number of reads at this codon for that sample.</p>
<p><cite>icodon</cite> : the number of the codon (or residue) for which we
are plotting the amino acid frequencies. Based on 1, 2, ...
numbering. Each gene in <cite>counts</cite>  must have a codon with
this number.</p>
<p><cite>plotfile</cite> : the name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p><cite>aas</cite> : a list of the amino acids for which we are plotting the
the fraction. Must give one-letter amino acid codes, or &#8216;STOP&#8217;
for stop codons. Typically you might want this to be something
like all non-wildtype amino acids at the position. In the legend,
a &#8216;*&#8217; character is used to represent stop codons.</p>
<p><cite>title</cite> : string placed above the plot as a title.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
if these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotAlignmentStatistics">
<tt class="descclassname">plot.</tt><tt class="descname">PlotAlignmentStatistics</tt><big>(</big><em>infiles</em>, <em>names</em>, <em>plotfile</em><big>)</big><a class="headerlink" href="#plot.PlotAlignmentStatistics" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes plot of alignment statistics for one or more samples.</p>
<p><cite>infiles</cite> : a list of file name(s) containing the alignment
statistics. These should be in the format written by
<cite>mapmuts.io.WriteAlignmentStatistics</cite>.</p>
<p><cite>names</cite> : a list of strings giving the names of the samples
corresponding to each entry in <cite>infiles</cite>.</p>
<p><cite>plotfile</cite> : the name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt> to plot summary statistics
showing how the reads fall into different categories based on
whether they could be aligned. The data is taken from the files
in <cite>infiles</cite>. If there is just one infile (one sample), then a pie
chart is plotted. If there are multiple infiles (multiple
samples) then a stacked bar graph is plotted. The categories of
reads are:</p>
<blockquote>
<div><p><cite>filtered</cite> : equal to <cite>nfiltered</cite></p>
<p><cite>low Q</cite> : equal to <cite>nlowq</cite></p>
<p><cite>excess N</cite> : equal to <cite>nexcessn</cite></p>
<p><cite>unpaired</cite> : equal to <cite>nattempted - npaired</cite></p>
<p><cite>unaligned</cite> : equal to <cite>npaired - nalignedfullgene</cite></p>
<p><cite>outside gene</cite> : equal to <cite>nalignedfullgene - nalignedgene</cite></p>
<p><cite>aligned</cite> : equal to <cite>nalignedgene</cite></p>
</div></blockquote>
<p>This function requires <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
if these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotCodonDepth">
<tt class="descclassname">plot.</tt><tt class="descname">PlotCodonDepth</tt><big>(</big><em>infile</em>, <em>plotfile</em><big>)</big><a class="headerlink" href="#plot.PlotCodonDepth" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots codon read depth as a function of gene position.</p>
<p><cite>infile</cite> : name of input file of the type written by 
<cite>mapmuts.io.WriteCodonCounts</cite>.</p>
<p><cite>plotfile</cite> : name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt> to plot the read depth
as a function of codon position along a gene. Codon
positions are numbered 1, 2, ... from the start of the gene.
Plots number of times each codon is called by both paired,
a single paired read, and is called to an ambiguous identity.</p>
<p>This function requires <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotCodonDist">
<tt class="descclassname">plot.</tt><tt class="descname">PlotCodonDist</tt><big>(</big><em>infile</em>, <em>plotfile</em>, <em>dist_type</em><big>)</big><a class="headerlink" href="#plot.PlotCodonDist" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots distribution of types of codon mutations across gene.</p>
<p>Takes as input a file (<em>infile</em>) specifying the number of counts
for each codon along the gene. Creates a PDF plot (<em>plotfile</em>)
showing the frequency of different types of codon mutations
along the gene, as specified by <em>dist_type</em>. Only codon counts
for which all three nucleotides are definitively called (upper
case nucleotides in <em>infile</em>) are counted for the plots.</p>
<p><em>infile</em> : name of input file of the type written by 
<em>mapmuts.io.WriteCodonCounts</em>. This file specifies the counts
for all of the codon mutations at each position.</p>
<p><em>plotfile</em> : name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p><em>dist_type</em> : the type of distribution that is plotted along
the gene. It is a string with the following possible values:</p>
<blockquote>
<div><ul class="simple">
<li><em>syn_ns_stop</em> specifies that we plot the distribution
of synonymous, nonsynonymous, and stop codons.</li>
<li><em>codon_nmuts</em> specifies that we plot the distribution
of the number of nucleotide mutations per codon for 
all codon mutations (1, 2, or 3).</li>
</ul>
</div></blockquote>
<p>This function requires <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotCorrelation">
<tt class="descclassname">plot.</tt><tt class="descname">PlotCorrelation</tt><big>(</big><em>xs</em>, <em>ys</em>, <em>plotfile</em>, <em>xlabel</em>, <em>ylabel</em>, <em>logx=False</em>, <em>logy=False</em>, <em>corr=None</em>, <em>title=False</em>, <em>alpha=1.0</em>, <em>symmetrize=False</em>, <em>fixaxes=False</em>, <em>additionalxy=</em>, <span class="optional">[</span><span class="optional">]</span><em>bigmargin=0.26</em>, <em>xsize=1.8</em><big>)</big><a class="headerlink" href="#plot.PlotCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the correlation between two variables as a scatter plot.</p>
<p>The data is plotted as a scatter plot.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an Exception if
these modules cannot be imported (if <em>PylabAvailable() == False</em>).</p>
<p>The calling variables use LaTex format for strings. So for
example, &#8216;$10^5$&#8217; will print the LaTex equivalent of this 
string. Similarly, certain raw text strings (such as those
including underscores) will cause problems if you do not
escape the LaTex format meaning. For instance, &#8216;x_label&#8217;
will cause a problem since underscore is not valid
outside of math mode in LaTex, so you would need to use
&#8216;x_label&#8217; to escape the underscore.</p>
<p>CALLING VARIABLES:</p>
<ul class="simple">
<li><em>xs</em> and <em>ys</em> are lists of numbers, with the lists
being of the same length. Entry <em>xs[i]</em> is plotted
on the x-axis against entries <em>ys[i]</em> on the y-axis.</li>
<li><em>plotfile</em> is a string giving the name of the plot PDF file
that we create. It should end in the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.
If this plot already exists, it is overwritten.</li>
<li><em>xlabel</em> is a string giving the label placed on the x-axis.</li>
<li><em>ylabel</em> is a string giving the label placed on the y-axis.</li>
<li><em>logx</em> specifies that we log transform the data in <em>xs</em>.
This is <em>False</em> by default; set to <em>True</em> if you want
to log transform (base 10 logarithms) the data.</li>
<li><em>logy</em> is like <em>logx</em>, but for the data in <em>ys</em>.</li>
<li><em>corr</em> specifies if we calculate and include a correlation
coefficient on the plot. If it is <em>None</em>, then no
correlation is computed. Otherwise, the coefficient
is calculated using <tt class="docutils literal"><span class="pre">scipy</span></tt> (so this requires
<em>mapmuts.bayesian.ScipyAvailable() == True</em>). In this case, <em>corr</em> should
be set to the string <em>Pearson</em> (to calculate the 
Pearson linear correlation coefficient) or to the string
<em>Spearman</em> (to calculate Spearman&#8217;s rho rank-order correlation).
If <em>logx</em> and / or <em>logy</em> are true, then the data are log
transformed BEFORE calculating the correlation, so the coefficient
is for correlation between the log-transformed data.
In both cases, the correlations are reported along with the
two-tailed P-values. They are written on the plot.
If you set this to a value other than <em>None</em> and <tt class="docutils literal"><span class="pre">scipy</span></tt> is
not available, then an exception is raised.
If <em>additionalxy</em> specifies additional data, this data is
combined with that in <em>xs</em> and <em>ys</em> to calculate the correlation.</li>
<li><em>title</em> is a string giving the title placed above the plot. 
It can be <em>False</em> if no title is to be used. Otherwise, it should
be the title string (using LaTex formatting, spaces are allowed).
Is <em>False</em> by default.</li>
<li><em>alpha</em> is the transparency of the plotted points. By default
it is one, which means that there is no transparency. If you make
the value closer to zero, then the points become partially transparent.
The rationale is that if you have many overlapping points, making
them partially transparent helps you better see the density.
At any position on the plot, the intensity will be saturated
when there are 1.0 / alpha points plotted. So a reasonable value
of <em>alpha</em> might be something like 0.1.</li>
<li><em>symmetrize</em> is an optional argument that is <em>False</em> by default.
If <em>True</em>, we make the X and Y limits on the plot the same.</li>
<li><em>fixaxes</em> is an optional argument that is <em>False</em> by default.
If <em>True</em>, we fix both the X and Y axes to go from 0 to 1, 
with ticks at 0, 0.5, and 1. If you set this option to <em>True</em>,
then you must set <em>logx</em> and <em>logy</em> to <em>False</em>.</li>
<li><em>additionalxy</em> is an optional list. By default, it is an empty list.
However, you can use it to plot additional data points with a different
color. The main data (specified by <em>xs</em> and <em>ys</em>) is plotted with
blue circles. For each addition set of data that you want to plot,
include a 2-tuple of lists of the form <em>(x2s, y2s)</em> in 
<em>additionalxy</em>. Currently, only one such 2-tuple is allowed
(so only one additional data set). These are plotted as red
triangles, whereas the main data is plotted as blue
circles. The same <em>alpha</em> value set by the <em>alpha</em> parameter applies
to these points as well. The data in <em>additionalxy</em> is combined
with that in <em>xs</em> and <em>ys</em> to calculate the correlation
if <em>corr</em> is being used.</li>
<li><em>bigmargin</em> is an optional argument that is 0.26 by default. It is
the fraction of the plot width taken up by the larger margin, which
is the bottom and left. Make larger if you need more space for axis
labels.</li>
<li><em>xsize</em> is an optional argument that is the width of the plot in inches.
It is 1.8 by default.</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="plot.PlotEnrichmentRatios">
<tt class="descclassname">plot.</tt><tt class="descname">PlotEnrichmentRatios</tt><big>(</big><em>mutations</em>, <em>ratios</em>, <em>ratio_errs</em>, <em>plotfile</em><big>)</big><a class="headerlink" href="#plot.PlotEnrichmentRatios" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots enrichment ratios for different mutations on a log scale.</p>
<p><em>mutations</em> : a list of strings giving the names of the mutations,
such as <em>[&#8216;M1A&#8217;, &#8216;M1C&#8217;]</em>.</p>
<p><em>ratios</em> : a list of the same lengths as <em>mutations</em> giving
the enrichment ratios.</p>
<p><em>ratio_errs</em> : a list containing two lists each 
of the same length as mutations giving
the upper and lower error bar bounds at 2-tuple entries. Set this 
to <em>None</em> if you do not want to plot any error bars.</p>
<p><em>plotfile</em> : the name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
if these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotEquilibriumFreqs">
<tt class="descclassname">plot.</tt><tt class="descname">PlotEquilibriumFreqs</tt><big>(</big><em>pis</em>, <em>plotfile</em>, <em>title=''</em>, <em>pi_errs=None</em>, <em>differentialpreferences=False</em><big>)</big><a class="headerlink" href="#plot.PlotEquilibriumFreqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots equilibrium frequencies of different amino acids.</p>
<p><em>pis</em> : a dictionary keyed by amino acid codes with values giving
the equilibrium frequency.</p>
<p><em>plotfile</em> : the name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p><em>title</em> is an optional argument giving the string for the title affixed
to the top of the plot. Is an empty string (&#8216;&#8217;) by default.</p>
<p><em>pi_errs</em> is an optional argument that is <em>None</em> by default. If it is
set to another value, it should specify the error bars for the plot.
In this case, for each key in <em>pis</em>, there should be a similar key
in <em>pi_errs</em> with the value a list or tuple such that <em>pi_errs[key][0]</em>
gives the lower bound and <em>pi_errs[key][1]</em> gives the upper bound.</p>
<p><em>differentialpreferences</em> is an optional argument that is <em>False</em> by
default. Set to <em>True</em> if you are plotting differential preferences
rather than preferences.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
if these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotInsertLengths">
<tt class="descclassname">plot.</tt><tt class="descname">PlotInsertLengths</tt><big>(</big><em>infile</em>, <em>plotfile</em>, <em>normed=True</em><big>)</big><a class="headerlink" href="#plot.PlotInsertLengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots distribution of insert lengths.</p>
<p><cite>infile</cite> : name of input file containing the distribution of insert
lengths, in the format written by <cite>mapmuts.io.WriteInsertLengths</cite>.</p>
<p><cite>plotfile</cite> : name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). Must end in the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p><cite>normed</cite> : Boolean switch specifying that we plot the fractional
distribution of insert lengths (so that all bars sum to one).
This is done if this switch has its default value of <cite>True</cite>. If
set to <cite>False</cite>, instead plot the actual number of reads of each
length.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt> to make a histogram showing
the distribution of insert lengths.</p>
<p>This function requires <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
if these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotLinearDensity">
<tt class="descclassname">plot.</tt><tt class="descname">PlotLinearDensity</tt><big>(</big><em>datalist</em>, <em>plotfile</em>, <em>xlabel</em>, <em>ylabel</em>, <em>title=False</em>, <em>fixymax=False</em><big>)</big><a class="headerlink" href="#plot.PlotLinearDensity" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots linear density of variable as a function of primary sequence.</p>
<p>This function is designed to plot some variable (such as the number
of epitopes as a function of the primary sequence position). It
creates an output PDF plot <em>plotfile</em>.</p>
<p>The data is plotted as lines. If there is more than one
data series to be plotted, a legend is included.</p>
<p>This function uses pylab / matplotlib. It will raise an Exception if
these modules cannot be imported (if <em>PylabAvailable() == False</em>).</p>
<p>The calling variables use LaTex format for strings. So for
example, &#8216;$10^5$&#8217; will print the LaTex equivalent of this 
string. Similarly, certain raw text strings (such as those
including underscores) will cause problems if you do not
escape the LaTex format meaning. For instance, &#8216;x_label&#8217;
will cause a problem since underscore is not valid
outside of math mode in LaTex, so you would need to use
&#8216;x_label&#8217; to escape the underscore.</p>
<p>CALLING VARIABLES:</p>
<ul class="simple">
<li><em>datalist</em>  is a list specifying the data to plot. It should
be a list of one or more 2-tuples of the form 
<em>(label, data)</em> where <em>label</em> is a string label used in
the legend, and data is a list of 2-tuples <em>(x, y)</em>
specifying the points to be plotted.</li>
<li><em>plotfile</em> is a string giving the name of the plot PDF file
that we create. It should end in the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.
If this plot already exists, it is overwritten.</li>
<li><em>xlabel</em> is a string giving the label placed on the x-axis.</li>
<li><em>ylabel</em> is a string giving the label placed on the y-axis.</li>
<li><em>title</em> is a string giving the title placed above the plot. 
It can be <em>False</em> if no title is to be used. Otherwise, it should
be the title string (using LaTex formatting, spaces are allowed).
Is <em>False</em> by default.</li>
<li><em>fixymax</em> means that we fix the y-maximum to the specified value.
This may be useful if you are making multiple plots for comparisons
between them, and want them all to have the same y-maximum.
Note that the value specified here is taken to be the data
maximum &#8211; the actually maximum of the y-axis is somewhat
higher to provide some padding space. Is <em>False</em> by default.</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="plot.PlotMutCountFracs">
<tt class="descclassname">plot.</tt><tt class="descname">PlotMutCountFracs</tt><big>(</big><em>plotfile</em>, <em>title</em>, <em>names</em>, <em>all_cumulfracs</em>, <em>syn_cumulfracs</em>, <em>all_counts</em>, <em>syn_counts</em>, <em>legendloc</em>, <em>writecounts=True</em><big>)</big><a class="headerlink" href="#plot.PlotMutCountFracs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots fraction of mutations with &gt;= a given number of counts.</p>
<p>Does this for both all mutations and synonymous mutations. The plots
are placed side by side.</p>
<p>The plots show the fractions of mutations
that are found &gt;= <em>n</em> times for a range of values of <em>n</em>.</p>
<p>Requires <em>pylab</em> / <em>matplotlib</em>, and will raise an exception
if not available for import.</p>
<p>CALLING VARIABLES:</p>
<ul>
<li><p class="first"><em>plotfile</em> : string giving name of the created plot file. Must end in
the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
</li>
<li><p class="first"><em>title</em> : string giving the title placed about the plot.</p>
</li>
<li><p class="first"><em>names</em> : a list of strings giving the names of the samples to
plot.</p>
</li>
<li><p class="first"><em>all_cumulfracs</em> : a list of the same length as <em>names</em> giving
the cumulative fractions for all mutations. Each entry should be a list, 
and all of these lists should be the same length. <em>cumulfracs[i][n]</em>
gives the fraction of mutations to sample <em>names[i]</em> that
are found &gt;= <em>n</em> times. The x-axis of the created plot
will go from 0 to <em>len(all_cumulfracs[0]) - 1</em>.</p>
</li>
<li><p class="first"><em>syn_cumulfracs</em> : a list like <em>all_cumulfracs</em> except for synonymous
mutations.</p>
</li>
<li><p class="first"><em>all_counts</em> : integer counts of all mutations (the total number of mutations
used for <em>all_cumulfracs</em>), used to create plot title.</p>
</li>
<li><p class="first"><em>syn_counts</em> : like <em>all_counts</em> but for synonymous mutations.</p>
</li>
<li><p class="first"><em>legendloc</em> : specifies the location of the legend. Should be a string.
Valid values are:</p>
<blockquote>
<div><ul class="simple">
<li><em>bottom</em> : put legend at the bottom of the plot.</li>
<li><em>right</em> : put legend at the right of the plot.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>writecounts</em> is a Boolean switch specifying whether we include the counts of all
mutations (specified by <em>all_counts</em> and <em>syn_counts</em>) in the plot title. We do
this if <em>writecounts</em> is <em>True</em>, and do not if it is <em>False</em>.</p>
</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="plot.PlotMutFracs">
<tt class="descclassname">plot.</tt><tt class="descname">PlotMutFracs</tt><big>(</big><em>infiles</em>, <em>names</em>, <em>plotfile</em>, <em>keys</em>, <em>writefracs</em>, <em>formatting={}</em>, <em>textwritefracs=None</em><big>)</big><a class="headerlink" href="#plot.PlotMutFracs" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes plot of mutation fractions per codon for several samples.</p>
<p>For each sample, calculates the fraction of all codon counts
(called by both reads) that represent the types of mutations 
specified by <em>keys</em>. Makes a stacked bar graph showing these
counts. The fractions are for all sites in the gene
aggregated together.</p>
<p><cite>infiles</cite> : a list of file names containing the codon counts for
each sample, in the format written by <cite>mapmuts.io.WriteCodonCounts</cite>.
Note that there is also an option for each entry in this list to
be a dictionary rather than a file &#8211; in this case, the dictionary 
should already contain the information that would normally be
read from the file. Specifically, it must contain a key
for each of the strings in <em>keys</em> giving the fraction of sites
with that type of mutation as a value.</p>
<p><cite>names</cite> : a list of strings giving the names of the samples
corresponding to each entry in <cite>infiles</cite>.</p>
<p><cite>plotfile</cite> : name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p><cite>keys</cite> : a list specifying the order (from bottom to top) that we
stack the bars. It can contain any of the following entries:</p>
<blockquote>
<div><ul class="simple">
<li>&#8216;synonymous&#8217; : fraction of synonymous mutations.</li>
<li>&#8216;nonsynonymous&#8217; : fraction of nonsynonymous mutations.</li>
<li>&#8216;stop codon&#8217; : fraction of stop codon mutations.</li>
<li>&#8216;1 nucleotide mutation&#8217; : fraction of mutations with one
nucleotide change.</li>
<li>&#8216;2 nucleotide mutations&#8217; : fraction of mutations with two
nucleotide changes.</li>
<li>&#8216;3 nucleotide mutations&#8217; : fraction of mutations with three
nucleotide changes.</li>
</ul>
</div></blockquote>
<p>Most typically you would want to set <em>keys</em> to either
<em>[&#8216;synonymous&#8217;, &#8216;nonsynonymous&#8217;, &#8216;stop codon&#8217;]</em> or to
<em>[&#8216;1 nucleotide mutation&#8217;, &#8216;2 nucleotide mutations&#8217;, &#8216;3 nucleotide mutations&#8217;]</em>.</p>
<p><cite>writefracs</cite> is a Boolean switch specifying whether we report the
numerical fraction of mutations in each category above the bar. If
<em>True</em> then we report these fractions; if <em>False</em> then we do not.</p>
<p><em>formatting</em> is a dictionary in which you can use string keys and their
values to specify additional formatting. Specifically:</p>
<blockquote>
<div><ul class="simple">
<li><em>figwidth</em> specifies the width of the figure to be this
in inches rather than the function&#8217;s default width of 5.75.</li>
</ul>
</div></blockquote>
<ul class="simple">
<li><em>textwritefracs</em> is an option that allows you to write the fraction
of mutations in each category to a file in text format. By default,
this option is <em>None</em>, meaning that no such writing is done. If you
want to do such writing, set it to a string representing a filename.
This file is then created and will list the fractions of mutations
in each category in <em>keys</em> for each of the samples in <em>names</em>.</li>
</ul>
<p>This function requires <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
if these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotNTMutFracs">
<tt class="descclassname">plot.</tt><tt class="descname">PlotNTMutFracs</tt><big>(</big><em>infiles</em>, <em>names</em>, <em>plotfile</em><big>)</big><a class="headerlink" href="#plot.PlotNTMutFracs" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes plot of nucleotide mutation types for several samples.</p>
<p>For each sample, calculates the fraction of each nucleotide call
(called by both reads) that represents each type of nucleotide
substitution (i.e. C-&gt;G, G-&gt;C; A-&gt;G, C-&gt;T); etc). Makes a stacked
bar graph showing the fraction of sites with these mutations
over aggregated over all positions.</p>
<p><cite>infiles</cite> : a list of file names containing the nucleotide counts for
each sample, in the format written by <cite>mapmuts.io.WriteNTCounts</cite>.
Note that there is also an option for each entry in this list to
be a dictionary rather than a file &#8211; in this case, the dictionary 
should already contain the information that would normally be
read from the file. Specifically, it must contain a key
for each entry in <cite>sequtils.NTMutTypes()</cite> giving the fraction of sites
with that type of mutation as a value.</p>
<p><cite>names</cite> : a list of strings giving the names of the samples
corresponding to each entry in <cite>infiles</cite>.</p>
<p><cite>plotfile</cite> : the name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
if these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotPairedMutFracs">
<tt class="descclassname">plot.</tt><tt class="descname">PlotPairedMutFracs</tt><big>(</big><em>infiles</em>, <em>names</em>, <em>plotfile</em><big>)</big><a class="headerlink" href="#plot.PlotPairedMutFracs" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes paired bar graph of mutation fractions per codon.</p>
<p>For each sample, calculates the fraction of all codon counts
(called by both reads) that represent synonymous / nonsynoymous /
stop codons, and one- / two- / three-nucleotide mutations. Makes
paired stacked bar graphs showing these fractions. The fractions
for all sites in the gene are aggregated together.</p>
<p><cite>infiles</cite> : a list of file names containing the codon counts for
each sample, in the format written by <cite>mapmuts.io.WriteCodonCounts</cite>.
Note that there is also an option for each entry in this list to
be a dictionary rather than a file &#8211; in this case, the dictionary 
should already contain the information that would normally be
read from the file. Specifically, it must contain a key
for each of the strings in <em>keys</em> giving the fraction of sites
with that type of mutation as a value.</p>
<p><cite>names</cite> : a list of strings giving the names of the samples
corresponding to each entry in <cite>infiles</cite>.</p>
<p><cite>plotfile</cite> : name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p>Requires <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>, will raise an exception
if these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotReadMismatches">
<tt class="descclassname">plot.</tt><tt class="descname">PlotReadMismatches</tt><big>(</big><em>infile</em>, <em>plotfile</em>, <em>logy</em>, <em>title=None</em><big>)</big><a class="headerlink" href="#plot.PlotReadMismatches" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots distribution of read mismatches.</p>
<p><cite>infile</cite> : name of input file containing the distribution of
mismatches along the length of a read, in the format written
by <cite>mapmuts.io.WriteReadMismatches</cite>.</p>
<p><cite>plotfile</cite> : name of the output plot file created by this method
(such as &#8216;plot.pdf&#8217;). The extension must be <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</p>
<p><cite>logy</cite> : Boolean switch specifying that we use a log scale for the
y-axis (fraction of mismatches). We use a log scale if and only
if this switch is <cite>True</cite>.</p>
<p><cite>title</cite> : string specifying a title placed above the plot. Can be
set to <cite>None</cite> if you do not want a title.</p>
<p>This function uses <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt> to plot the distribution
of mismatches along the length of the gene. The fraction of reads with
a mutation at a given site are shown in 1, 2, ... numbering from the
start of the read.</p>
<p>This function requires <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt>. It will raise an exception
if these modules cannot be imported (if <cite>PylabAvailable() == False</cite>).</p>
</dd></dl>

<dl class="function">
<dt id="plot.PlotTraces">
<tt class="descclassname">plot.</tt><tt class="descname">PlotTraces</tt><big>(</big><em>traces</em>, <em>plotfile</em>, <em>xlabel</em>, <em>ylabel</em>, <em>title</em>, <em>trace_labels=None</em><big>)</big><a class="headerlink" href="#plot.PlotTraces" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots traces from an MCMC run.</p>
<p>Requires <tt class="docutils literal"><span class="pre">pylab</span></tt> and will raise an exception of <em>PylabAvailable()</em>
is <em>False</em>.</p>
<p>CALLING VARIABLES:</p>
<ul class="simple">
<li><em>traces</em> should be a list of one or more lists or arrays. All
entries must be the same length. These should give the traces
of the posterior to be plotted (it is up to you if you want
to have applied burn-ins or thinning before the traces). In the
legend, if <em>trace_labels</em> is <em>None</em>, then the different traces 
are labeled as #1, #2, etc. Otherwise <em>trace_labels</em> should be 
a list of strings of the same length as <em>traces</em> that gives
the labels applied to the traces.</li>
<li><em>plotfile</em> : name of the PDF file created by this method.
Should have the extension <tt class="docutils literal"><span class="pre">.pdf</span></tt>.</li>
<li><em>xlabel</em> : string giving the label on the X-axis.</li>
<li><em>ylabel</em> : string giving the label on the Y-axis.</li>
<li><em>title</em> : string giving the title placed above the plot.</li>
<li><em>trace_labels</em> : an optional argument that is <em>None</em> by default.
If set to some other value, it should be a list of strings of the
same length as <em>traces</em>. The strings are used to label the corresponding
traces in <em>traces</em> in the legend. Furthermore, if an identical label
appears multiple times in <em>trace_labels</em>, it is only shown on the 
legend once. So if you have multiple runs of the same sample in
<em>traces</em>, just give them the same label in <em>trace_labels</em> and then they 
will be shown with the same linestyle and only one legend label.</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="plot.PylabAvailable">
<tt class="descclassname">plot.</tt><tt class="descname">PylabAvailable</tt><big>(</big><big>)</big><a class="headerlink" href="#plot.PylabAvailable" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <cite>True</cite> if <tt class="docutils literal"><span class="pre">pylab</span></tt> / <tt class="docutils literal"><span class="pre">matplotlib</span></tt> available, <cite>False</cite> otherwise.</p>
<p>You should call this function to test for the availability of the
<tt class="docutils literal"><span class="pre">pylab</span></tt> and <tt class="docutils literal"><span class="pre">matplotlib</span></tt> plotting modules before using other functions in
this module.</p>
</dd></dl>

<dl class="function">
<dt id="plot.SplitLabel">
<tt class="descclassname">plot.</tt><tt class="descname">SplitLabel</tt><big>(</big><em>label</em>, <em>splitlen</em>, <em>splitchar</em><big>)</big><a class="headerlink" href="#plot.SplitLabel" title="Permalink to this definition">¶</a></dt>
<dd><p>Splits a string with a return if it exceeds a certain length.</p>
<p><cite>label</cite> : a string giving the label we might split.</p>
<p><cite>splitlen</cite> : the maximum length of a label before we attempt to 
split it.</p>
<p><cite>splitchar</cite> : the character added when splitting a label.</p>
<p>If <cite>len(label) &gt; splitlen</cite>, we attempt to split the label in the
middle by adding <cite>splitchar</cite>. The label is split as close to the
middle as possible while splitting at a space.</p>
<p>EXAMPLES:</p>
<p>No splitting as label length less than <cite>splitlen</cite>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">SplitLabel</span><span class="p">(</span><span class="s">&#39;WT virus 1&#39;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="go">&#39;WT virus 1&#39;</span>
</pre></div>
</div>
<p>Splitting of this label:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">SplitLabel</span><span class="p">(</span><span class="s">&#39;WT plasmid 1&#39;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="go">&#39;WT\nplasmid 1&#39;</span>
</pre></div>
</div>
<p>Splitting of this label:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">SplitLabel</span><span class="p">(</span><span class="s">&#39;mutated WT plasmid 1&#39;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="go">&#39;mutated WT\nplasmid 1&#39;</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><tt class="docutils literal"><span class="pre">plot</span></tt> Module</a><ul>
<li><a class="reference internal" href="#list-of-functions">List of functions</a></li>
<li><a class="reference internal" href="#documentation-for-functions">Documentation for functions</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="simulate_module.html"
                        title="previous chapter"><tt class="docutils literal"><span class="pre">simulate</span></tt> Module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="latex_module.html"
                        title="next chapter"><tt class="docutils literal"><span class="pre">latex</span></tt> Module</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/plot_module.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="latex_module.html" title="latex Module"
             >next</a> |</li>
        <li class="right" >
          <a href="simulate_module.html" title="simulate Module"
             >previous</a> |</li>
        <li><a href="index.html">mapmuts 1.1 documentation</a> &raquo;</li>
          <li><a href="pythonapi.html" >Python modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2013, Jesse Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>