============
mapmuts
============

This package is obsolete!
-----------------------------

**IMPORTANT NOTE:** `mapmuts`_ has been superseded by `dms_tools`_. Unless you need legacy functions of `mapmuts`_, we suggest that you use `dms_tools`_ instead.

Package information
----------------------

`mapmuts`_ is a `Python`_ package for analyzing deep-sequencing data to identify mutation frequencies in mutant libraries of a gene. It is specifically designed for the analysis of overlapping paired-end reads.

`mapmuts`_ source code is available `on GitHub`_. 

`mapmuts documentation`_ is available `on GitHub Pages`_.

`mapmuts`_ was written by `Jesse Bloom`_.

`mapmuts`_ is free software under the `GPLv3`_ license. 

If you use `mapmuts`_, then `please cite mapmuts as described here`_. 


.. _`on GitHub`: https://github.com/jbloom/mapmuts
.. _`on GitHub Pages`: http://jbloom.github.io/mapmuts/
.. _`mapmuts documentation`: http://jbloom.github.io/mapmuts/
.. _`Jesse Bloom`: http://research.fhcrc.org/bloom/en.html
.. _`mapmuts`: https://github.com/jbloom/mapmuts
.. _`Python`: http://www.python.org/
.. _`GPLv3`: http://www.gnu.org/licenses/gpl.html
.. _`please cite mapmuts as described here`: http://jbloom.github.io/mapmuts/acknowledgements.html#citations
.. _`dms_tools`: http://jbloomlab.github.io/dms_tools/
